---
title: "GSE157775 MIDGUT Elife"
output: html_notebook
---

#Objective: 

## 1. Check if there is inflammaging on a holistic level (day 60)
## 2. Check if lab's Bulk IMs are expressed in other cells in midgut

#Important files***:
## 1. GSE157775_cell_annotation.csv: Auhtor's annotation
## 2. GSE157775_sce.RDS: RDS of Seurat OBJ the author's put on GEO
*** These files can be downloaded at https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157775

################################
set wk dir and library packages
```{r}
setwd(getwd())
getwd()
library(Seurat)
library(ggplot2)
library(cowplot)
library(Matrix)
library(dplyr)
library(data.table)
library(tidyverse)
library(remotes)
library(data.table)
library(ggrepel)
library(DESeq2)
library(openxlsx)
```

load saved wkimage if needed
```{r}
load("./MIDGUT_Elife_wkimage.RData")
```


create an obj of III markers (lab's Bulk IMs)
```{r}
III_Markers <- c("SPH93",
                 "CecA1",
                 "CecA2",
                 "AttA",
                 "AttB",
                 "AttC",
                 "DptA",
                 "DptB",
                 "Dro",
                 "Mtk",
                 "IM18",
                 "edin")
```


Import SOBJ and author's annotation
```{r}
Author_metadata <- fread("./raw_data/GSE157775_cell_annotation.csv",GUTer = T)
MIDGUT_Elife <- readRDS("./raw_data/GSE157775_sce.RDS")

#convert into seurat OBJ
MIDGUT_Elife <- as.Seurat(MIDGUT_Elife)
```

First check if cell IDs match
```{r}
table(
  MIDGUT_Elife@meta.data$cellID%in%Author_metadata$cellID
)
```
Export the seurat object to a dataframe where rows are gene IDs, and columns are cells; Need to convert FBgn to gene symbols first with flybase ID validator, then convert the datafram back to SOBJ (otherwise features generated in every slot will be FBgn, currently cannot be directly converted, not even coerce)
```{r}
expr_df <- GetAssayData(MIDGUT_Elife, layer = "counts")  # for counts

# Convert to a dataframe
expr_df <- as.data.frame(as.matrix(expr_df))

# Ensure the row names are gene IDs and column names are cell IDs
write.table(file = "./processed_data/SOBJ/MIDGUT_SOBJ_FBgn.txt",
            rownames(expr_df),
            sep = ",",
            quote = F,
            col.names = F,
            row.names = F
            )

```


Provides tables for converting and annotating Ensembl Gene IDs.
```{r}
#devtools::install_github("stephenturner/annotables")
library(annotables)
BDGP6 <- annotables::bdgp6

table(duplicated(BDGP6$symbol))
table(duplicated(rownames(expr_df)))

BDGP6 <- BDGP6[!duplicated(BDGP6$ensgene),]
BDGP6 <- BDGP6[!duplicated(BDGP6$symbol),]
table(duplicated(BDGP6$symbol))


table(rownames(expr_df)%in%BDGP6$ensgene)
```


Convert FBgn to Gene symbol
```{r}
convert_fbgn_to_symbol <- function(expr_df, annotation_df) {
  expr_df$FBgn <- rownames(expr_df)

  # Merge annotation (BDGP6) with expression data
  merged_df <- merge(expr_df, annotation_df[, c("ensgene", "symbol")], 
                     by.x = "FBgn", by.y = "ensgene", all.x = TRUE)

  # Ensure unique gene symbols
  merged_df$symbol <- make.unique(as.character(merged_df$symbol))

  # Set rownames as FBgn (to avoid duplicates)
  rownames(merged_df) <- merged_df$FBgn

  # Keep Gene Symbol as a separate column
  merged_df <- merged_df[, c("symbol", setdiff(colnames(merged_df), c("symbol", "FBgn")))]

  return(merged_df)
}



# Convert FBgn IDs to Gene Symbols
converted_expr_df <- convert_fbgn_to_symbol(expr_df, BDGP6)
converted_expr_df <- na.omit(converted_expr_df)
rownames(converted_expr_df) <- converted_expr_df$symbol
converted_expr_df <- converted_expr_df[,-1]

dim(converted_expr_df)
```

convert mapped expr to SOBJ
```{r}
MIDGUT_Elife2 <- CreateSeuratObject(counts = converted_expr_df)

# add cellID in meta.data slot of SOBJ
MIDGUT_Elife2@meta.data$'cellID' <- colnames(converted_expr_df)
```


For the scRNA-seq experiments, we used the wild-type WDah fly line and dissected whole midguts at young (~7d), mid-age (~30d) and old time-points (~60d)

Whole midguts from adult female flies were dissected in 1xPBS and fixed in a fixative solution 

Drosophila melanogaster reference genome (BDGP6)

format metadata; merge author's annotation with Seurat OBJ
```{r}
merge_metadata <- function(seurat_obj, external_metadata, cell_id_col) {
  # Extract the metadata from the Seurat object
  seurat_metadata <- seurat_obj@meta.data
  
  # Merge the Seurat metadata with the external metadata
  merged_metadata <- merge(
    seurat_metadata, 
    external_metadata, 
    by = cell_id_col, 
    all.x = TRUE
  )
  
  # Ensure the rownames of the merged metadata match the original Seurat object
  rownames(merged_metadata) <- merged_metadata[[cell_id_col]]
  
  # Replace the Seurat object's metadata with the merged metadata
  seurat_obj@meta.data <- merged_metadata
  
  return(seurat_obj)
}

# Example usage:
MIDGUT_Elife2 <- merge_metadata(MIDGUT_Elife2, Author_metadata,cell_id_col = "cellID")

```

change some names of meta.data slot so functions from 57KBrain dataset can be used if needed ("Genotype","sex","Age","Replicate")
Note: this midgut dataset has only: In total, we had two young, one mid-age and two old samples.
```{r}
#View(MIDGUT_Elife2@meta.data)
# batch is replicate; age is age; assigned_cell_type is annotation; 
MIDGUT_Elife2@meta.data$'Genotype' <- "WDah_WT"
MIDGUT_Elife2@meta.data$'sex' <- "Mix"
colnames(MIDGUT_Elife2@meta.data)
colnames(MIDGUT_Elife2@meta.data)[15] <- "Replicate"
colnames(MIDGUT_Elife2@meta.data)[16] <- "Age"
colnames(MIDGUT_Elife2@meta.data)[20] <- "annotation"


#look at how many replicates of each age exist in this dataset
table(paste0(MIDGUT_Elife2@meta.data$Age,"_",MIDGUT_Elife2@meta.data$replicate))


```

change active ident to assigned_cell_type
```{r}
Idents(MIDGUT_Elife2) <- MIDGUT_Elife2@meta.data$annotation
```






#run standard pipeline of seurat
```{r}
# run standard anlaysis workflow
MIDGUT_Elife2 <- NormalizeData(MIDGUT_Elife2)
MIDGUT_Elife2 <- FindVariableFeatures(MIDGUT_Elife2)
MIDGUT_Elife2 <- ScaleData(MIDGUT_Elife2)

min(nrow(MIDGUT_Elife2), ncol(MIDGUT_Elife2))
MIDGUT_Elife2 <- RunPCA(MIDGUT_Elife2,npcs = 50,verbose = T)


MIDGUT_Elife2 <- FindNeighbors(MIDGUT_Elife2, dims = 1:30, reduction = "pca", verbose=TRUE)


MIDGUT_Elife2 <- FindClusters(MIDGUT_Elife2, resolution = 0.6, cluster.name = "MIDGUT_Elife_clusters")

```


#Run UMAP to build the reduciton slot of SOBJ
```{r}

MIDGUT_Elife3 <- RunUMAP(MIDGUT_Elife2, dims = 1:30, reduction = "pca", reduction.name = "UMAP_MIDGUT_Elife",verbose=T)



#change UMAP reduction coordinates to Author's

GUT(MIDGUT_Elife3@reductions[["UMAP_MIDGUT_Elife"]]@cell.embeddings)

#change reduciton UAMP coordinates to the one in authos's annotation
MIDGUT_Elife3@reductions[["UMAP_MIDGUT_Elife"]]@cell.embeddings[,1] <- MIDGUT_Elife3@meta.data$UMAP_1

MIDGUT_Elife3@reductions[["UMAP_MIDGUT_Elife"]]@cell.embeddings[,2] <- MIDGUT_Elife3@meta.data$UMAP_2

GUT(MIDGUT_Elife3@reductions[["UMAP_MIDGUT_Elife"]]@cell.embeddings)

#change active idents to annotation
Idents(MIDGUT_Elife3) <- MIDGUT_Elife3@meta.data$annotation
```


#DimPlot of UMAP
```{r}

#Now plot the UMAP Dimplot
pdf("./results/Dimplot/MIDGUT_Elife_UMAP_Dimplot.pdf",width = 9,height = 6)
DimPlot(MIDGUT_Elife3,reduction = "UMAP_MIDGUT_Elife",raster = F,label = T,group.by = "annotation")
dev.off()

```
Objective 1. Check if there is inflammaging on a holistic level (day 60)

aggregate expression, split by genotype, sex, age, and replicate.
```{r}
Pseudo_GUT_Cells <- as.data.frame(
  AggregateExpression(MIDGUT_Elife3,
                      assays = "RNA",
                      group.by = c("Genotype","sex","Age","Replicate")
                      )$RNA
  )

dim(Pseudo_GUT_Cells)
#[1] 12871     5

colnames(Pseudo_GUT_Cells) <- c("WDah_Mixed_30_1","WDah_Mixed_60_2","WDah_Mixed_60_3","WDah_Mixed_7_1","WDah_Mixed_7_2")
colnames(Pseudo_GUT_Cells)

#Export the the converted Pseudobulk expression to local dir for DE
openxlsx::write.xlsx(Pseudo_GUT_Cells,
                     "./processed_data/Pseudo_GUT_Cells.xlsx",
                     rowNames=T
                     )
```

Make a function to Run DESeq2 for Pseudo_GUT_Cells variable.

This function is dependent on DO_DEseq function in DEseq2_Differential_Expression_Feb22_2024.docx on OneDrive.

This function is solely written for objectice 1.1&1.3 to compare this GUT57k data with lab's Bulk_RNAseq data on a holistic level.

```{r}
#library the DE_DEseq function (see star_DEseq protocol on onedrive)
Do_DEseq <- 
  function(Counts_matrix,
           n_TR,                # 
           n_CR,                #
           p_thresh,            #
           log2FC_thresh,       #
           File_Title){        #
  # make meta for DEseq (make a variable that tells DEseq which columns are controls and which are treatment ones)
  
meta <- colnames(Counts_matrix)

meta <- as.data.frame(meta)

rownames(meta) <- meta[,1]

  meta$"Condition" <- c(rep("TR",n_TR),rep("CR",n_CR))
  
  
  meta$Condition <- factor(meta$Condition,levels = c("CR","TR"))
  
  # make the dds object for DEseq
  
  dds <- DESeqDataSetFromMatrix(countData = Counts_matrix,
                              colData = meta,
                              design= ~ Condition)
  dds <- DESeq(dds)
  res <- results(dds, pAdjustMethod = "BH")
  
  #extracting result
  res1 <- data.frame(res, stringsAsFactors = FALSE, check.names = FALSE)

  res1 <- na.omit(res1)
res1 <- res1[order(res1$padj, res1$log2FoldChange, decreasing = c(FALSE, TRUE)), ]


res1[which(res1$log2FoldChange >= log2FC_thresh & res1$pvalue < p_thresh),'sig'] <- 'up'
res1[which(res1$log2FoldChange <= -log2FC_thresh & res1$pvalue < p_thresh),'sig'] <- 'down'
res1[which(abs(res1$log2FoldChange) <= log2FC_thresh | res1$pvalue >= p_thresh),'sig'] <- 'none'


print(
  paste0(File_Title," up ","total DEGs: ",nrow(res1[res1$'sig'=='up',]))
      )
print(res1[res1$'sig'=='up',]
      )

print(
  paste0(File_Title," down ","total DEGs: ",nrow(res1[res1$'sig'=='down',]))
      )
print(res1[res1$'sig'=='down',]
      )


write.xlsx(res1, paste0("./processed_data/DESeq_result/",File_Title,".xlsx"),rowNames=T)

write.xlsx(res1[res1$'sig'=='up',],paste0("./processed_data/DESeq_result/",File_Title,"_UP",p_thresh,".xlsx"),rowNames=T)

write.xlsx(res1[res1$'sig'=='down',],paste0("./processed_data/DESeq_result/",File_Title,"_DOWN",p_thresh,".xlsx"),rowNames=T)


}


#This function is dependent on DO_DEseq function
Run_DESeq2_Pseudo_GUT_Cells <- function(PSEUDO_INPUT,STRAIN,SEX,OLD_AGE,YOUNG_AGE){
  #First remove genes containing just zero;
  #PSEUDO_INPUT <- PSEUDO_INPUT[rowSums(PSEUDO_INPUT)>1,];
  #PSEUDO_INPUT <- na.omit(PSEUDO_INPUT);
  #PSEUDO_INPUT <- PSEUDO_INPUT+1
  #First take the subset based on strain and sex
  PSEUDO_INPUT <- PSEUDO_INPUT[,which(str_split(colnames(PSEUDO_INPUT),pattern = "_",4,simplify =T)[,1]==STRAIN)];
  PSEUDO_INPUT <- PSEUDO_INPUT[,which(str_split(colnames(PSEUDO_INPUT),pattern = "_",4,simplify =T)[,2]==SEX)];
  #Obtain the index for old age and young age columns
  index_old <- which(str_split(colnames(PSEUDO_INPUT),pattern = "_",4,simplify =T)[,3]==OLD_AGE);
  index_young <- which(str_split(colnames(PSEUDO_INPUT),pattern = "_",4,simplify =T)[,3]==YOUNG_AGE);
  #Obtain the # of columns in each condition
  N_OLD <- length(index_old);
  N_YOUNG <- length(index_young);
  
  #Run DO_DEseq function to generate results to ./DESeq_result dir
  Do_DEseq(Counts_matrix=cbind(PSEUDO_INPUT[,index_old],PSEUDO_INPUT[,index_young]), #select the columns in readcount matrix and select treatment sample first then control samples
           n_TR=N_OLD,   #number of treatment sample             
           n_CR=N_YOUNG,    #number of control sample            
           p_thresh=0.05,         #p value threshold   
           log2FC_thresh=1,       #log2 fold change threshold
           File_Title=paste0("DESeq_",STRAIN,"_",SEX,"_",OLD_AGE,"_vs_",YOUNG_AGE)
           )  #xlsx file title

}                                              
```


Run_DESeq2_Pseudo_HEAD_Cells function Using Universally acknowledged thresholds: p<0.05 and log2FC>1 (or you can manually select based on P value in .xlsx that contains both up and down)
Some results will later be used for generating ranked gene lists of DGRP-551 immune genes for female and male
```{r}

Run_DESeq2_Pseudo_GUT_Cells(
  PSEUDO_INPUT=Pseudo_GUT_Cells,
  STRAIN="WDah",
  SEX="Mixed",
  OLD_AGE="30",
  YOUNG_AGE="7"
  )

Run_DESeq2_Pseudo_GUT_Cells(
  PSEUDO_INPUT=Pseudo_GUT_Cells,
  STRAIN="WDah",
  SEX="Mixed",
  OLD_AGE="60",
  YOUNG_AGE="7"
  )

```



#save wk image
```{r}
save.image("./MIDGUT_Elife_wkimage.RData")
```

