---
title: "R Notebook for GSE107451 57k dataset analysis"
output: NA
---
# this is the R script for the analysis of the head ScRNA-Seq data of the GSE107451 

# Objectives (Partial; Full objectives can be found in the old file GSE107451_ALL_HEAD57k): 
# 1.1 does the scRNA-Seq data agree with bulk RNA-Seq data? 
# 1.2 which cells in the brain express the IMs? Is it the hemocytes? 
# 1.3 which one shows age-dependent increase or decrease? (Based on 1.1, 1.2 result, in plasmatocytes)? 
# 5.1 Based on result of 1.1, which immune genes in GO_ImmuneDef.txt are up-regulated DEGs in this scRNA-Seq dataset on a pseudo bulk level? 
# 5.2 In which clusters are these genes expressed in? 

# Dataset
## GSE107451 https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107451 
## we download the GSE107451_DGRP-551_w1118_WholeBrain_57k_0d_1d_3d_6d_9d_15d_30d_50d_10X_DGEM_MEX.mtx.tsv.tar.gz 
## we used Author's annotation as metadata of this analysis: GSE107451_DGRP-551_w1118_WholeBrain_57k_Metadata.tsv.gz
## which was based on analysis down by the authors of the paper 
## A Single-Cell Transcriptome Atlas of the Aging Drosophila Brain. 
##  - Cell 2018 Aug 9;174(4):982-998.e20. PMID: 29909982
##  - https://pubmed.ncbi.nlm.nih.gov/29909982/
## For analysis of scRNAseq data, Seurat V5 is used. For more information, check https://satijalab.org/seurat/, and https://satijalab.org/seurat/articles/install.html for specific instruction on installation.

                         
# Most important environment variables:
## ALL_HEAD_57k3: the seurat object for this analysis after every piece of information is aggregated into it. All Seurat plots are beased on this object.

## Pseudo_HEAD_Cells: aggregated count split by genotype, sex, age, and replicate. DEseq was performed on the basis of this aggregated count matrix to compare with bulk RNAseq data.

## Pseudo_HEAD_Genotype_Sex_Age_Cluster: aggregated count split by genotype, sex, age, and, cell type.
## TPM_HEAD_Genotype_Sex_Age_Cluster: aggregated count (Pseudo_HEAD_Genotype_Sex_Age_Cluster) normalized to TPM. Used for visualization with F1, F2, F3, F4, and F5.


set working directory and library package
```{r}
## run at HiPerGator (HPG) if mem not enough
setwd(getwd())  
getwd()

library(Seurat)
library(ggplot2)
library(Matrix)
library(dplyr)
library(data.table)
library(tidyverse)
library(openxlsx)
library(ggrepel)
library(gplots)
library(DESeq2)

```

Load work image (if there is one)
```{r}
#This is the work.image saved on local terminal
load("GSE107451_head_ALL_57k_clean_padj0.01.RData")
```

Import data and create Seurat object
```{r}
#import barcode, matrix, and feature files obtained by cellranger under folder GSE107451_DGRP-551_w1118_WholeBrain_57k_0d_1d_3d_6d_9d_15d_30d_50d_10X_DGEM_MEX.mtx.tsv
ALL_HEAD_57k <- Seurat::Read10X("./raw_data/GSE107451_DGRP-551_w1118_WholeBrain_57k_0d_1d_3d_6d_9d_15d_30d_50d_10X_DGEM_MEX.mtx.tsv")

#create seurat object, setting minimum standard (57k is filtered dataset by author, no need to filter)
ALL_HEAD_57k <- CreateSeuratObject(counts = ALL_HEAD_57k,
                               project = "ALL_HEAD_57k_GSE107451", 
                               min.cells = 0,  
                               min.features = 0)
#look at the meta data
View(ALL_HEAD_57k@meta.data)

#import metadata of author's and add that to Seurat OBJ
meta_author <- as.data.frame(fread("./raw_data/GSE107451_DGRP-551_w1118_WholeBrain_57k_Metadata.tsv",header = T))
#check number of cells in the 57k dataset
table(rownames(ALL_HEAD_57k@meta.data)==meta_author$new_barcode)
#TRUE 
#56902 

#adding new_barcode column in meta.data slot of Seurat OBJ from author's metadata
ALL_HEAD_57k@meta.data$"new_barcode" <- rownames(ALL_HEAD_57k@meta.data)
#merge author's metadata to the meta.data slot of Seurat OBJ
ALL_HEAD_57k@meta.data <- cbind(ALL_HEAD_57k@meta.data,meta_author)

#look at the meta.data after adding author's meta data
View(ALL_HEAD_57k@meta.data)
```

Normalization and find clusters under different resolutions
```{r}
###NOTE: Running the standard workflow from normalization, pca, etc.. to tSNE is simply for adding the "reduction" slot in the Seurat OBJ; The authors' annotations will be used in this analysis
ALL_HEAD_57k2 <- NormalizeData(ALL_HEAD_57k)
ALL_HEAD_57k2 <- FindVariableFeatures(ALL_HEAD_57k2)
ALL_HEAD_57k2 <- ScaleData(ALL_HEAD_57k2)
min(nrow(ALL_HEAD_57k2), ncol(ALL_HEAD_57k2))
ALL_HEAD_57k2 <- RunPCA(ALL_HEAD_57k2,npcs = 20,verbose = F)
ALL_HEAD_57k2 <- FindNeighbors(ALL_HEAD_57k2, dims = 1:20, reduction = "pca", verbose=TRUE)

ALL_HEAD_57k2@active.ident <- as.factor(ALL_HEAD_57k2@meta.data$annotation)
```

Run tSNE reductons
```{r}
###NOTE: Running the standard workflow tSNE is simply for adding the "tSNE" slot in the Seurat OBJ

#Run tSNE
ALL_HEAD_57k3 <- RunTSNE(ALL_HEAD_57k2, dims = 1:20, reduction = "pca", reduction.name = "tSNE_ALL_HEAD_57k_Author_cluster",verbose=T,perplexity=30)

 
#change tSNE coordinates to authors' values in the annotation (so the topology looks the same)
table(rownames(ALL_HEAD_57k3@meta.data)%in%rownames(ALL_HEAD_57k3@reductions[["tSNE_ALL_HEAD_57k_Author_cluster"]]@cell.embeddings))
# TRUE 
# 56902 
ALL_HEAD_57k3@reductions[["tSNE_ALL_HEAD_57k_Author_cluster"]]@cell.embeddings[,1] <- ALL_HEAD_57k3@meta.data$seurat_tsne1
ALL_HEAD_57k3@reductions[["tSNE_ALL_HEAD_57k_Author_cluster"]]@cell.embeddings[,2] <- ALL_HEAD_57k3@meta.data$seurat_tsne2

#In seuratV5, joinlayers is needed to prevent error when running FindallMarkers because data slot location has changed since V3
ALL_HEAD_57k3 <- JoinLayers(ALL_HEAD_57k3)
```

Create a Gene List of lab's Bulk inflammaging markers (IMs) and corresponding feature plots
```{r}
#creating inflammaging marker gene list (lab's Bulk IMs)
III_Marker_Genes <- c("SPH93",
                      "CecA1",
                      "CecA2",
                      "AttA",
                      "AttB",
                      "AttC",
                      "DptA",
                      "DptB",
                      "Dro",
                      "Mtk",
                      "IM18",
                      "edin")
```

Objective 1.1&1.3 to compare this HEAD57k data with lab's Bulk_RNAseq data on a holistic level. Which increased or decreased
Generate pseudobulk dataframe (rownames genes, colnames cells), split by genotype, sex, age, and replicate.
```{r}

Pseudo_HEAD_Cells <- as.data.frame(
  AggregateExpression(ALL_HEAD_57k3,
                      assays = "RNA",
                      group.by = c("Genotype","sex","Age","Replicate")
                      )$RNA
  )

dim(Pseudo_HEAD_Cells)
#17473    52

#Export the the converted Pseudobulk expression to local dir for DE
openxlsx::write.xlsx(Pseudo_HEAD_Cells,
                     "./processed_data/Pseudo_HEAD_Cells_57k.xlsx",
                     rowNames=T
                     )
```

Make a function to Run DESeq2 for Pseudo_HEAD_Cells variable.

This function is dependent on DO_DEseq function in DEseq2_Differential_Expression_Feb22_2024.docx on OneDrive.

This function is solely written for objectice 1.1&1.3 to compare this HEAD57k data with lab's Bulk_RNAseq data on a holistic level.
For instance, comparisons like day 50 vs day 3 for DGRP-551 female.


```{r}
#library the DE_DEseq function (see star_DEseq protocol on onedrive)
Do_DEseq <- 
  function(Counts_matrix,
           n_TR,                # 
           n_CR,                #
           p_thresh,            #
           log2FC_thresh,       #
           File_Title){        #
  # make meta for DEseq (make a variable that tells DEseq which columns are controls and which are treatment ones)
  
meta <- colnames(Counts_matrix)

meta <- as.data.frame(meta)

rownames(meta) <- meta[,1]

  meta$"Condition" <- c(rep("TR",n_TR),rep("CR",n_CR))
  
  
  meta$Condition <- factor(meta$Condition,levels = c("CR","TR"))
  
  # make the dds object for DEseq
  
  dds <- DESeqDataSetFromMatrix(countData = Counts_matrix,
                              colData = meta,
                              design= ~ Condition)
  dds <- DESeq(dds)
  res <- results(dds, pAdjustMethod = "BH")
  
  #extracting result
  res1 <- data.frame(res, stringsAsFactors = FALSE, check.names = FALSE)

  res1 <- na.omit(res1)
res1 <- res1[order(res1$padj, res1$log2FoldChange, decreasing = c(FALSE, TRUE)), ]


res1[which(res1$log2FoldChange >= log2FC_thresh & res1$padj < p_thresh),'sig'] <- 'up'
res1[which(res1$log2FoldChange <= -log2FC_thresh & res1$padj < p_thresh),'sig'] <- 'down'
res1[which(abs(res1$log2FoldChange) <= log2FC_thresh | res1$padj >= p_thresh),'sig'] <- 'none'


print(
  paste0(File_Title," up ","total DEGs: ",nrow(res1[res1$'sig'=='up',]))
      )
print(res1[res1$'sig'=='up',]
      )

print(
  paste0(File_Title," down ","total DEGs: ",nrow(res1[res1$'sig'=='down',]))
      )
print(res1[res1$'sig'=='down',]
      )


write.xlsx(res1, paste0("./processed_data/DESeq_result/",File_Title,".xlsx"),rowNames=T)

write.xlsx(res1[res1$'sig'=='up',],paste0("./processed_data/DESeq_result/",File_Title,"_UP",p_thresh,".xlsx"),rowNames=T)

write.xlsx(res1[res1$'sig'=='down',],paste0("./processed_data/DESeq_result/",File_Title,"_DOWN",p_thresh,".xlsx"),rowNames=T)


}


#This function is dependent on DO_DEseq function
# This second function selects columns needed for DESeq and then runs Do_DESeq
Run_DESeq2_Pseudo_HEAD_Cells <- function(PSEUDO_INPUT,STRAIN,SEX,OLD_AGE,YOUNG_AGE){
  #First remove genes containing just zero;
  #PSEUDO_INPUT <- PSEUDO_INPUT[rowSums(PSEUDO_INPUT)>1,];
  #PSEUDO_INPUT <- na.omit(PSEUDO_INPUT);
  #PSEUDO_INPUT <- PSEUDO_INPUT+1
  #First take the subset based on strain and sex
  PSEUDO_INPUT <- PSEUDO_INPUT[,which(str_split(colnames(PSEUDO_INPUT),pattern = "_",4,simplify =T)[,1]==STRAIN)];
  PSEUDO_INPUT <- PSEUDO_INPUT[,which(str_split(colnames(PSEUDO_INPUT),pattern = "_",4,simplify =T)[,2]==SEX)];
  #Obtain the index for old age and young age columns
  index_old <- which(str_split(colnames(PSEUDO_INPUT),pattern = "_",4,simplify =T)[,3]==OLD_AGE);
  index_young <- which(str_split(colnames(PSEUDO_INPUT),pattern = "_",4,simplify =T)[,3]==YOUNG_AGE);
  #Obtain the # of columns in each condition
  N_OLD <- length(index_old); # number of replicate in old age
  N_YOUNG <- length(index_young); # number of replicate in young age
  
  #Run DO_DEseq function to generate results to ./DESeq_result dir
  Do_DEseq(Counts_matrix=cbind(PSEUDO_INPUT[,index_old],PSEUDO_INPUT[,index_young]), #select the columns in readcount matrix and select treatment sample first then control samples
           n_TR=N_OLD,   #number of treatment sample             
           n_CR=N_YOUNG,    #number of control sample            
           p_thresh=0.01,         #padj threshold   
           log2FC_thresh=1,       #log2 fold change threshold
           File_Title=paste0("DESeq_",STRAIN,"_",SEX,"_",OLD_AGE,"_vs_",YOUNG_AGE)
           )  #xlsx file title

}                                              
```

Run_DESeq2_Pseudo_HEAD_Cells function Using Universally acknowledged thresholds: p<0.05 and log2FC>1 (or you can manually select based on P value in .xlsx that contains both up and down)
Some results will later be used for generating ranked gene lists of DGRP-551 immune genes for female and male

```{r}

Run_DESeq2_Pseudo_HEAD_Cells(
  PSEUDO_INPUT=Pseudo_HEAD_Cells,
  STRAIN="DGRP-551",
  SEX="female",
  OLD_AGE="30",
  YOUNG_AGE="3"
  )

Run_DESeq2_Pseudo_HEAD_Cells(
  PSEUDO_INPUT=Pseudo_HEAD_Cells,
  STRAIN="DGRP-551",
  SEX="female",
  OLD_AGE="50",
  YOUNG_AGE="3"
  )

Run_DESeq2_Pseudo_HEAD_Cells(
  PSEUDO_INPUT=Pseudo_HEAD_Cells,
  STRAIN="DGRP-551",
  SEX="male",
  OLD_AGE="30",
  YOUNG_AGE="3"
  )

Run_DESeq2_Pseudo_HEAD_Cells(
  PSEUDO_INPUT=Pseudo_HEAD_Cells,
  STRAIN="DGRP-551",
  SEX="male",
  OLD_AGE="50",
  YOUNG_AGE="3"
  )

Run_DESeq2_Pseudo_HEAD_Cells(
  PSEUDO_INPUT=Pseudo_HEAD_Cells,
  STRAIN="w1118",
  SEX="female",
  OLD_AGE="30",
  YOUNG_AGE="3"
  )


Run_DESeq2_Pseudo_HEAD_Cells(
  PSEUDO_INPUT=Pseudo_HEAD_Cells,
  STRAIN="w1118",
  SEX="male",
  OLD_AGE="30",
  YOUNG_AGE="3"
  )
```



Import DESeq result of DGRP551 female and male
```{r}
DESeq_DGRP551_female_50_vs_3 <- read.xlsx("./processed_data/DESeq_result/DESeq_DGRP-551_female_50_vs_3.xlsx",
                                                  rowNames = T)
DESeq_DGRP551_male_50_vs_3 <- read.xlsx("./processed_data/DESeq_result/DESeq_DGRP-551_male_50_vs_3.xlsx",
                                                  rowNames = T)
DESeq_DGRP551_female_30_vs_3 <- read.xlsx("./processed_data/DESeq_result/DESeq_DGRP-551_female_30_vs_3.xlsx",
                                                  rowNames = T)
DESeq_DGRP551_male_30_vs_3 <- read.xlsx("./processed_data/DESeq_result/DESeq_DGRP-551_male_30_vs_3.xlsx",
                                                  rowNames = T)
```

Plot volcano plot of DESeq results
```{r}

plot_volcano <- function(res, log2FC_cutoff = 1, pval_cutoff = 0.05, 
                         xlim_range = c(-5, 5), ylim_range = c(0,100), label_top_n = 10, plot_title = "Volcano Plot") {
  library(ggplot2)
library(dplyr)
library(ggrepel)
  # Check if required columns exist
  required_cols <- c("log2FoldChange", "pvalue", "padj")
  missing_cols <- setdiff(required_cols, colnames(res))
  if (length(missing_cols) > 0) {
    stop(paste("Dataframe is missing required columns:", paste(missing_cols, collapse = ", ")))
  }

  # If 'Gene' column is missing, try using row names
  if (!"Gene" %in% colnames(res)) {
    if (!is.null(rownames(res))) {
      res$Gene <- rownames(res)
    } else {
      stop("No gene symbols found. Ensure gene names are in row names or a 'Gene' column.")
    }
  }

  # Define significance categories
  res$significance <- "none"
  res$significance[res$log2FoldChange >= log2FC_cutoff & res$padj < pval_cutoff] <- "up"
  res$significance[res$log2FoldChange <= -log2FC_cutoff & res$padj < pval_cutoff] <- "down"
  res$significance <- factor(res$significance, levels = c("none", "up", "down"))

  # Select top DEGs for labeling
  top_genes <- res %>%
    filter(padj < pval_cutoff) %>%
    arrange(padj) %>%
    head(label_top_n)

  # Generate Volcano Plot
  ggplot(res, aes(x = log2FoldChange, y = -log10(pvalue), color = significance)) +
    geom_point(alpha = 1, size = 2) +
    scale_color_manual(values = c("none" = "grey", "up" = "red", "down" = "blue")) +
    geom_vline(xintercept = c(-log2FC_cutoff, log2FC_cutoff), linetype = "dashed") +
    geom_hline(yintercept = -log10(pval_cutoff), linetype = "dashed") +
    xlim(xlim_range) +
    ylim(ylim_range) +
    labs(title = plot_title, x = "Log2 Fold Change", y = "-Log10 P-value") +
    theme_minimal() +
    geom_text_repel(data = top_genes, aes(label = Gene), 
                    size = 4, box.padding = 0.5, point.padding = 0.3, 
                    segment.color = "black",max.overlaps = Inf,
                    show.legend = FALSE) +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),  # Center title
    plot.title.position = "plot"  # Ensure it's at the top
  )
}


#DESeq_DGRP551_female_50_vs_3
plot_volcano(res=DESeq_DGRP551_female_50_vs_3,
             log2FC_cutoff = 1,
             pval_cutoff = 0.01, 
             xlim_range = c(-12, 12), 
             ylim_range=c(0,75), 
             label_top_n = 25, 
             plot_title = "DESeq_DGRP551_female_50_vs_3"
             )

#DESeq_DGRP551_male_50_vs_3
plot_volcano(res=DESeq_DGRP551_male_50_vs_3,
             log2FC_cutoff = 1,
             pval_cutoff = 0.01, 
             xlim_range = c(-8, 12), 
             ylim_range=c(0,40), 
             label_top_n = 25, 
             plot_title = "DESeq_DGRP551_male_50_vs_3"
             )
```


Objective 5.1 Based on result of 1.1, which immune genes in GO_ImmuneDef.txt are up-regulated DEGs in this scRNA-Seq dataset on a pseudo bulk level? 

Import GO_ImmuneDef
```{r}
GO_ImmuneDef <- fread("./GO_ImmuneDef.txt",header = F)
```

Make a ranked gene list based on GO_ImmuneDef.txt and DESeq results sorted by padj.  
```{r}
View(
  subset(DESeq_DGRP551_female_50_vs_3,sig=="up"&rownames(DESeq_DGRP551_female_50_vs_3)%in%GO_ImmuneDef$V1)
  )

View(
  subset(DESeq_DGRP551_male_50_vs_3,sig=="up"&rownames(DESeq_DGRP551_male_50_vs_3)%in%GO_ImmuneDef$V1)
  )

View(
  subset(DESeq_DGRP551_female_30_vs_3,sig=="up"&rownames(DESeq_DGRP551_female_30_vs_3)%in%GO_ImmuneDef$V1)
  )

View(
  subset(DESeq_DGRP551_male_30_vs_3,sig=="up"&rownames(DESeq_DGRP551_male_30_vs_3)%in%GO_ImmuneDef$V1)
  )




write.xlsx(subset(DESeq_DGRP551_female_50_vs_3,sig=="up"&rownames(DESeq_DGRP551_female_50_vs_3)%in%GO_ImmuneDef$V1),
           rowNames=T,
           "./processed_data/ranked_ImmuneGene_List/DGRP551_female_50_vs_3_ImmuneGene_List.xlsx")

write.xlsx(subset(DESeq_DGRP551_male_50_vs_3,sig=="up"&rownames(DESeq_DGRP551_male_50_vs_3)%in%GO_ImmuneDef$V1),
           rowNames=T,
           "./processed_data/ranked_ImmuneGene_List/DGRP551_male_50_vs_3_ImmuneGene_List.xlsx")


write.xlsx(subset(DESeq_DGRP551_female_30_vs_3,sig=="up"&rownames(DESeq_DGRP551_female_30_vs_3)%in%GO_ImmuneDef$V1),
           rowNames=T,
           "./processed_data/ranked_ImmuneGene_List/DGRP551_female_30_vs_3_ImmuneGene_List.xlsx")

write.xlsx(subset(DESeq_DGRP551_male_30_vs_3,sig=="up"&rownames(DESeq_DGRP551_male_30_vs_3)%in%GO_ImmuneDef$V1),
           rowNames=T,
           "./processed_data/ranked_ImmuneGene_List/DGRP551_male_30_vs_3_ImmuneGene_List.xlsx") 
```


Objective 5.2 In which clusters are these genes expressed in? 

First join target immune genes based on the two xlsx files exported above
```{r}
#
female_target_ImmuneGenes_50 <- 
  rownames(
    subset(DESeq_DGRP551_female_50_vs_3,sig=="up"&rownames(DESeq_DGRP551_female_50_vs_3)%in%GO_ImmuneDef$V1)
    )
female_target_ImmuneGenes_50
#
male_target_ImmuneGenes_50 <- 
  rownames(
    subset(DESeq_DGRP551_male_50_vs_3,sig=="up"&rownames(DESeq_DGRP551_male_50_vs_3)%in%GO_ImmuneDef$V1)
    )
male_target_ImmuneGenes_50
#
female_target_ImmuneGenes_30 <- 
  rownames(
    subset(DESeq_DGRP551_female_30_vs_3,sig=="up"&rownames(DESeq_DGRP551_female_30_vs_3)%in%GO_ImmuneDef$V1)
    )
female_target_ImmuneGenes_30
#
male_target_ImmuneGenes_30 <- 
  rownames(
    subset(DESeq_DGRP551_male_30_vs_3,sig=="up"&rownames(DESeq_DGRP551_male_30_vs_3)%in%GO_ImmuneDef$V1)
    )
male_target_ImmuneGenes_30
```

Based on findallmarkers result, determine the percentage expression of these immune genes in different clusters. Potential clusters can be identified.

Find annotated clusters first
```{r}
sort(unique(ALL_HEAD_57k3$annotation))

annotated_clusters <- unique(ALL_HEAD_57k3$annotation)[
   !grepl("^[0-9]+$", unique(ALL_HEAD_57k3$annotation))
   ]
sort(annotated_clusters)
```

generate tSNE only for annotated clusters
```{r}
pdf(file = "./results/Dimplot/annotated_clusters_tSNE.pdf",width = 5,height = 5)
DimPlot(subset(ALL_HEAD_57k3, idents = annotated_clusters),
        reduction = "tSNE_ALL_HEAD_57k_Author_cluster",
        group.by = "annotation",
        label = T,
        label.size = 2,
        repel = TRUE)+ NoLegend()
dev.off()
```


make a function for dotplot
```{r}
Make_dotplot <- function(SOBJ, GENELIST,FILE_TITLE) {
  Idents(SOBJ) <- SOBJ$annotation;
  annotated_clusters <- unique(SOBJ$annotation)[
  !grepl("^[0-9]+$", unique(SOBJ$annotation))
  ];
  #all clusters
  Var_dotplot <- DotPlot(SOBJ, features = GENELIST,group.by = "annotation")+
    theme(axis.text.x = element_text(angle = 90, size = 12), legend.text = element_text(size = 12), legend.title = element_text(size = 12))+
  coord_flip()+ scale_x_discrete(limits = rev(GENELIST))  # Force the order of genes;
  
  #export dotplot of female_target_ImmuneGenes_50
pdf(paste0("./results/Dotplot/",FILE_TITLE,".pdf"),
    width = 25,
    height = 6)
print(Var_dotplot)
dev.off()

# Make another dot plot that only includes annotated clusters
Var_dotplot_annotated <- DotPlot(subset(SOBJ, idents = annotated_clusters), features = GENELIST,group.by = "annotation")+
    theme(axis.text.x = element_text(angle = 90, size = 8), axis.text.y = element_text(size = 8), legend.text = element_text(size = 8), legend.title = element_text(size = 8))+
  coord_flip()+ scale_x_discrete(limits = rev(GENELIST))  # Force the order of genes

# save annotated plot
pdf(paste0("./results/Dotplot/",FILE_TITLE,"_annotated.pdf"),
    width = 15,
    height = 5)
print(Var_dotplot_annotated)
dev.off()
}
```



female 50
```{r}
Make_dotplot(SOBJ = ALL_HEAD_57k3, 
             GENELIST = female_target_ImmuneGenes_50,
             FILE_TITLE = "female_target_ImmuneGenes_50_dotplot"
             )
```


female 30
```{r}
Make_dotplot(SOBJ = ALL_HEAD_57k3, 
             GENELIST = female_target_ImmuneGenes_30,
             FILE_TITLE = "female_target_ImmuneGenes_30_dotplot"
             )
```

male 50
```{r}
Make_dotplot(SOBJ = ALL_HEAD_57k3, 
             GENELIST = male_target_ImmuneGenes_50,
             FILE_TITLE = "male_target_ImmuneGenes_50_dotplot"
             )
```


male 30
```{r}
Make_dotplot(SOBJ = ALL_HEAD_57k3, 
             GENELIST = male_target_ImmuneGenes_30,
             FILE_TITLE = "male_target_ImmuneGenes_30_dotplot"
             )
```

WB IMs
```{r}
Make_dotplot(SOBJ = ALL_HEAD_57k3, 
             GENELIST = III_Marker_Genes,
             FILE_TITLE = "WB_IMs_dotplot"
             )
```

dotplot of female_target_ImmuneGenes_50 and male_target_ImmuneGenes_50
```{r}
Make_dotplot(SOBJ = ALL_HEAD_57k3, 
             GENELIST = unique(c(female_target_ImmuneGenes_50,male_target_ImmuneGenes_50)),
             FILE_TITLE = "female_AND_male_target_ImmuneGenes_50_dotplot"
             )
```


dotplot of female_target_ImmuneGenes_50 and male_target_ImmuneGenes_50 in female DGRP-551 and male DGRP-551 (Combined gene lists in female DGRP-551 and male DGRP-551)
```{r}
subset(ALL_HEAD_57k3,sex=="female"&Genotype=="DGRP-551")
subset(ALL_HEAD_57k3,sex=="male"&Genotype=="DGRP-551")



Make_dotplot(SOBJ = subset(ALL_HEAD_57k3,sex=="female"&Genotype=="DGRP-551"),
             GENELIST = unique(c(female_target_ImmuneGenes_50,male_target_ImmuneGenes_50)),
             FILE_TITLE = "female_AND_male_IMs_50_in_DGRP-551_female_dotplot"
             )

Make_dotplot(SOBJ = subset(ALL_HEAD_57k3,sex=="male"&Genotype=="DGRP-551"),
             GENELIST = unique(c(female_target_ImmuneGenes_50,male_target_ImmuneGenes_50)),
             FILE_TITLE = "female_AND_male_IMs_50_in_DGRP-551_male_dotplot"
             )
```

dot plot that include only specific clusters of brain IMs
```{r}
Dotplot_IM_specific <- function(SOBJ, GENELIST,FILE_TITLE, CELLTYPE) {
  Idents(SOBJ) <- SOBJ$annotation;
  #all clusters
  Var_dotplot <- DotPlot(SOBJ, features = GENELIST,group.by = "annotation")+
    theme(axis.text.x = element_text(angle = 90, size = 12), legend.text = element_text(size = 12), legend.title = element_text(size = 12))+
  coord_flip()+ scale_x_discrete(limits = rev(GENELIST))  # Force the order of genes;
  
  #export dotplot of female_target_ImmuneGenes_50
pdf(paste0("./results/Dotplot/",FILE_TITLE,".pdf"),
    width = 25,
    height = 6)
print(Var_dotplot)
dev.off()

# Make another dot plot that only includes IM-specific celltypes
Var_dotplot_IM_celltype <- DotPlot(subset(SOBJ, idents = CELLTYPE), features = GENELIST,group.by = "annotation")+
    theme(axis.text.x = element_text(angle = 90, size = 8), axis.text.y = element_text(size = 8), legend.text = element_text(size = 8), legend.title = element_text(size = 8))+
  coord_flip()+ scale_x_discrete(limits = rev(GENELIST))  # Force the order of genes

# save IM-specific_celltypes plot
pdf(paste0("./results/Dotplot/",FILE_TITLE,"_IM-specific_celltypes.pdf"),
    width = 10,
    height = 5)
print(Var_dotplot_IM_celltype)
dev.off()
}



levels(subset(ALL_HEAD_57k3,sex=="female"&Genotype=="DGRP-551"))

# Female_brain_celltype
Female_brain_celltype <- c("Plasmatocytes", "Proc/Ms", "MBON", "Dopaminergic", "Astrocyte-like", 
                "PAM", "Tyraminergic", "Pm1/Pm2", "Perineurial_glia", "Photoreceptors", 
                "T2", "AstA/Nplp1", "L4/L5", "Serotonergic", "Ensheathing_glia", 
                "Subperineurial_glia", "Cortex_glia", "Tm5ab")

# female DGRP-551
Dotplot_IM_specific(SOBJ = subset(ALL_HEAD_57k3,sex=="female"&Genotype=="DGRP-551"),
                    GENELIST = unique(c(female_target_ImmuneGenes_50,male_target_ImmuneGenes_50)),
                    FILE_TITLE = "female_AND_male_IMs_50_in_DGRP-551_female_dotplot",
                    CELLTYPE = Female_brain_celltype
                    )






levels(subset(ALL_HEAD_57k3,sex=="male"&Genotype=="DGRP-551"))

# male_brain_celltype
male_brain_celltype <- c("MBON", "Plasmatocytes", "PAM", "Serotonergic", "Gr43a", 
                   "Olfactory_projection_neurons", "Subperineurial_glia", "Pm1/Pm2/Pm3", 
                   "Photoreceptors", "Tm5c", "L4/L5", "Pm1/Pm2", "Mi1", "L2", 
                   "adPN/C15&kn", "Astrocyte-like", "G-KC", "Perineurial_glia", 
                   "TmY14", "Dm8/Dm11", "Chiasm_glia", "Ensheathing_glia", 
                   "adPN/kn", "Lawf1", "T2", "Mip", "Octopaminergic", "C3", "Tm9"
                   )

# male DGRP-551
Dotplot_IM_specific(SOBJ = subset(ALL_HEAD_57k3,sex=="male"&Genotype=="DGRP-551"),
                    GENELIST = unique(c(female_target_ImmuneGenes_50,male_target_ImmuneGenes_50)),
                    FILE_TITLE = "female_AND_male_IMs_50_in_DGRP-551_male_dotplot",
                    CELLTYPE = male_brain_celltype
                    )


```




In addition to dotplot, straight forward demonstration of aggregated expression based on strain, sex, age and cluster could also be possible.
The visualization functions (in later parts) take 'Pseudo_HEAD_Genotype_Sex_Age_Cluster' variable as one input
```{r}
Pseudo_HEAD_Genotype_Sex_Age_Cluster <- as.data.frame(
  AggregateExpression(ALL_HEAD_57k3,
                      assays = "RNA",
                      group.by = c("Genotype","sex","Age","annotation")
                      )$RNA
  )

dim(Pseudo_HEAD_Genotype_Sex_Age_Cluster)
#[1] 17473  2845
#export the count file to ./large_file (not synced to github)

write.xlsx(Pseudo_HEAD_Genotype_Sex_Age_Cluster,"./large_file/Pseudo_HEAD_Genotype_Sex_Age_Cluster.xlsx",rowNames=T)
```


#import gtf file
```{r}
library(data.table)
genes <- fread("./large_file/dmel-all-r6.16.gtf",header = F)
setnames(genes, names(genes), c("chr","source","type","start","end","score","strand","phase","attributes") )

# extract gene info in gtf
genes <- genes[type == "gene"]

extract_attributes <- function(gtf_attributes, att_of_interest){
  att <- strsplit(gtf_attributes, "; ")
  att <- gsub("\"","",unlist(att))
  if(!is.null(unlist(strsplit(att[grep(att_of_interest, att)], " ")))){
    return( unlist(strsplit(att[grep(att_of_interest, att)], " "))[2])
  }else{
    return(NA)}
}

# this is how to, for example, extract the values for the attributes of interest (here: "gene_id")
genes$gene_id <- unlist(lapply(genes$attributes, extract_attributes, "gene_id"))
genes$gene_symbol <- unlist(lapply(genes$attributes, extract_attributes, "gene_symbol"))

# remove the semi-colon
genes$gene_symbol <-substr(genes$gene_symbol,1,nchar(genes$gene_symbol)-1)

genes$gene_length <- abs(genes$end-genes$start)

openxlsx::write.xlsx(genes,
           "./processed_data/transformed_gtf/dmel616gtf.xlsx",
           rowNames=F)
```

find joint genes in Pseudo_HEAD_Genotype_Sex_Age_Cluster and gtf file
```{r}
nrow(Pseudo_HEAD_Genotype_Sex_Age_Cluster)
nrow(genes)
nrow(genes[genes$gene_symbol%in%rownames(Pseudo_HEAD_Genotype_Sex_Age_Cluster),])

gene_SOBJ <- genes[genes$gene_symbol%in%rownames(Pseudo_HEAD_Genotype_Sex_Age_Cluster),]

table(
  gene_SOBJ$gene_symbol%in%rownames(Pseudo_HEAD_Genotype_Sex_Age_Cluster)
  )

write.xlsx(gene_SOBJ,"./processed_data/transformed_gtf/gene_SOBJ.xlsx")
```


convert the dataframe above into TPM based on code snippet from Michael Love (DESeq2 developer)
https://support.bioconductor.org/p/91218/#91256
```{r}
# Function to convert counts to TPM with estimated gene lengths
count2TPM <- 
  function(counts.mat,gene.length){
    x <- counts.mat / gene.length
    tpm.mat <- t( t(x) * 1e6 / colSums(x) )
    }

```

Import necessary files for TPM conversion
```{r}
#import gene_SOBJ
gene_SOBJ <- read.xlsx("./processed_data/transformed_gtf/gene_SOBJ.xlsx",colNames = T)
#import Pseudo_HEAD_Genotype_Sex_Age_Cluster
Pseudo_HEAD_Genotype_Sex_Age_Cluster <- read.xlsx("./large_file/Pseudo_HEAD_Genotype_Sex_Age_Cluster.xlsx",rowNames = T,colNames = T)

#check if rownames are the same
table(
  gene_SOBJ$gene_symbol%in%rownames(Pseudo_HEAD_Genotype_Sex_Age_Cluster)
)


# obatine ordered gene lengths
ordered_gene_lengths <- gene_SOBJ$gene_length[match(rownames(Pseudo_HEAD_Genotype_Sex_Age_Cluster), gene_SOBJ$gene_symbol)]
# Combine the target genes and their corresponding lengths into a new dataframe
result_df_SOBJ <- data.frame(gene_symbol = rownames(Pseudo_HEAD_Genotype_Sex_Age_Cluster), gene_length = ordered_gene_lengths)
```

Run count2TPM
```{r}
TPM_HEAD_Genotype_Sex_Age_Cluster <- 
  count2TPM(counts.mat = Pseudo_HEAD_Genotype_Sex_Age_Cluster,
            gene.length = result_df_SOBJ$gene_length
            )
TPM_HEAD_Genotype_Sex_Age_Cluster <- as.data.frame(
  TPM_HEAD_Genotype_Sex_Age_Cluster
)

#test colsum
sum2(TPM_HEAD_Genotype_Sex_Age_Cluster$`DGRP-551_female_0_0`)

#export the TPM file
write.xlsx(TPM_HEAD_Genotype_Sex_Age_Cluster,
           "./large_file/TPM_HEAD_Genotype_Sex_Age_Cluster.xlsx",
           rowNames=T
           )
```

Generate the second input: number of cells in each combinations of Genotype_Sex_Age_Cluster (split by meta.data: "Genotype","sex","Age","annotation")
```{r}
extract_cell_counts <- function(seurat_obj, metadata_order) {
  # Ensure column names in metadata are unique
  colnames(seurat_obj@meta.data) <- make.unique(colnames(seurat_obj@meta.data))
  
  # Check that metadata_order is valid
  if (!all(metadata_order %in% colnames(seurat_obj@meta.data))) {
    stop("Some elements of metadata_order are not present in the Seurat object's metadata.")
  }
  
  # Extract metadata and count cells in each unique combination
  metadata <- seurat_obj@meta.data
  metadata_grouped <- metadata %>%
    group_by(across(all_of(metadata_order))) %>%
    summarise(Cell_Count = n(), .groups = "drop")
  
  # Create column names in the format "Genotype_Sex_Age_Annotation"
  metadata_grouped <- metadata_grouped %>%
    mutate(Colname = paste(!!!syms(metadata_order), sep = "_"))
  
  # Create the output dataframe
  output_df <- as.data.frame(t(metadata_grouped$Cell_Count))
  colnames(output_df) <- metadata_grouped$Colname
  
  return(output_df)
}

#test to see if extraction is correct
nrow(subset(ALL_HEAD_57k3,annotation=="Plasmatocytes"&Age=="3"&Genotype=="DGRP-551"&sex=="female")@meta.data)
#[1] 20

#it is correct

# Usage example:
cell_number_Genotype_Sex_Age_Cluster <- extract_cell_counts(ALL_HEAD_57k3, c("Genotype", "sex", "Age", "annotation"))

#now export the information into processed data dir
write.xlsx(cell_number_Genotype_Sex_Age_Cluster,
           "./large_file/cell_number_Genotype_Sex_Age_Cluster.xlsx",
           rowNames=F)
```

F1 barplot function to show top clusters of a gene's expression at an age
```{r}

# function to create bar plots for a specific gene
plot_log2bars_Ages <- function(DATA, GENE, STRAIN, SEX, AGE1,TOP_N_CLUSTERS) {
  #first subset based on strain and sex
  DATA <- DATA[,which(str_split(colnames(DATA),pattern = "_",4,simplify =T)[,1]==STRAIN)];
  DATA <- DATA[,which(str_split(colnames(DATA),pattern = "_",4,simplify =T)[,2]==SEX)];
  
  #prepare data for two ages
  DATA_AGE1 <- DATA[,which(str_split(colnames(DATA),pattern = "_",4,simplify =T)[,3]==AGE1)];
  
  #log2 tranformation
  DATA_AGE1 <- log2(as.matrix(DATA_AGE1[GENE,]+1))
  #rank row-wise    
  DATA_AGE1 <- DATA_AGE1[, order(DATA_AGE1[1, ], decreasing = TRUE)];
  #select first 12 clusters to show
  DATA_AGE1 <- DATA_AGE1[1:TOP_N_CLUSTERS]
  DATA_AGE1 <- as.matrix(DATA_AGE1);
  DATA_AGE1 <- t(DATA_AGE1);
  
  barplot(DATA_AGE1,
          xlim=c(0,8), 
          ylim=c(0,20),
          width=0.4,
          main=paste0("log2(TPM)"," at day ",AGE1," of ", GENE," in ",STRAIN," ",SEX), 
          names.arg=str_split(colnames(DATA_AGE1),pattern = "_",4,simplify =T)[,4],
          ylab="log2Expression level", 
          xlab="Clusters",
          col=c("darkblue"),
          cex.names=0.4,
          las=2)
}

```


Run the function above (for example)
```{r}
#Run the function above

plot_log2bars_Ages (DATA =TPM_HEAD_Genotype_Sex_Age_Cluster,
                   STRAIN = "DGRP-551",
                   SEX = "female",
                   GENE ="Dro",
                   AGE1 ="50",
                   TOP_N_CLUSTERS = 12)

plot_log2bars_Ages (DATA =TPM_HEAD_Genotype_Sex_Age_Cluster,
                   STRAIN = "DGRP-551",
                   SEX = "female",
                   GENE ="Dro",
                   AGE1 ="30",
                   TOP_N_CLUSTERS = 12)
```

save F1 barplot in a batch
```{r}
#try saving in a batch (loop)

#female_target_ImmuneGenes_50 day 30
for (i in 1:length(female_target_ImmuneGenes_50)) {
  pdf(paste0("./results/Barplot/female_target_ImmuneGenes_50/Barplot_DGRP551_female_",female_target_ImmuneGenes_50[i],"_day30.pdf"),width = 9,height = 6)
plot_log2bars_Ages (DATA =TPM_HEAD_Genotype_Sex_Age_Cluster,
                   STRAIN = "DGRP-551",
                   SEX = "female",
                   GENE =female_target_ImmuneGenes_50[i],
                   AGE1 ="30",
                   TOP_N_CLUSTERS = 12)
dev.off()
}

#female_target_ImmuneGenes_50 day 50
for (i in 1:length(female_target_ImmuneGenes_50)) {
  pdf(paste0("./results/Barplot/female_target_ImmuneGenes_50/Barplot_DGRP551_female_",female_target_ImmuneGenes_50[i],"_day50.pdf"),width = 9,height = 6)
plot_log2bars_Ages (DATA =TPM_HEAD_Genotype_Sex_Age_Cluster,
                   STRAIN = "DGRP-551",
                   SEX = "female",
                   GENE =female_target_ImmuneGenes_50[i],
                   AGE1 ="50",
                   TOP_N_CLUSTERS = 12)
dev.off()
}


#male_target_ImmuneGenes_50 day 30
for (i in 1:length(male_target_ImmuneGenes_50)) {
  pdf(paste0("./results/Barplot/male_target_ImmuneGenes_50/Barplot_DGRP551_male_",male_target_ImmuneGenes_50[i],"_day30.pdf"),width = 9,height = 6)
plot_log2bars_Ages (DATA =TPM_HEAD_Genotype_Sex_Age_Cluster,
                   STRAIN = "DGRP-551",
                   SEX = "male",
                   GENE =male_target_ImmuneGenes_50[i],
                   AGE1 ="30",
                   TOP_N_CLUSTERS = 12)
dev.off()
}

#male_target_ImmuneGenes_50 day 50
for (i in 1:length(male_target_ImmuneGenes_50)) {
  pdf(paste0("./results/Barplot/male_target_ImmuneGenes_50/Barplot_DGRP551_male_",male_target_ImmuneGenes_50[i],"_day50.pdf"),width = 9,height = 6)
plot_log2bars_Ages (DATA =TPM_HEAD_Genotype_Sex_Age_Cluster,
                   STRAIN = "DGRP-551",
                   SEX = "male",
                   GENE =male_target_ImmuneGenes_50[i],
                   AGE1 ="50",
                   TOP_N_CLUSTERS = 12)
dev.off()
}


#female IMs day 30
for (i in 1:length(III_Marker_Genes)) {
  pdf(paste0("./results/Barplot/IMs/Barplot_DGRP551_female_",III_Marker_Genes[i],"_day30.pdf"),width = 9,height = 6)
plot_log2bars_Ages (DATA =TPM_HEAD_Genotype_Sex_Age_Cluster,
                   STRAIN = "DGRP-551",
                   SEX = "female",
                   GENE =III_Marker_Genes[i],
                   AGE1 ="30",
                   TOP_N_CLUSTERS = 12)
dev.off()
}

#female IMs day 50
for (i in 1:length(III_Marker_Genes)) {
  pdf(paste0("./results/Barplot/IMs/Barplot_DGRP551_female_",III_Marker_Genes[i],"_day50.pdf"),width = 9,height = 6)
plot_log2bars_Ages (DATA =TPM_HEAD_Genotype_Sex_Age_Cluster,
                   STRAIN = "DGRP-551",
                   SEX = "female",
                   GENE =III_Marker_Genes[i],
                   AGE1 ="50",
                   TOP_N_CLUSTERS = 12)
dev.off()
}


#male IMs day 30
for (i in 1:length(III_Marker_Genes)) {
  pdf(paste0("./results/Barplot/IMs/Barplot_DGRP551_male_",III_Marker_Genes[i],"_day30.pdf"),width = 9,height = 6)
plot_log2bars_Ages (DATA =TPM_HEAD_Genotype_Sex_Age_Cluster,
                   STRAIN = "DGRP-551",
                   SEX = "male",
                   GENE =III_Marker_Genes[i],
                   AGE1 ="30",
                   TOP_N_CLUSTERS = 12)
dev.off()
}

#male IMs day 50
for (i in 1:length(III_Marker_Genes)) {
  pdf(paste0("./results/Barplot/IMs/Barplot_DGRP551_male_",III_Marker_Genes[i],"_day50.pdf"),width = 9,height = 6)
plot_log2bars_Ages (DATA =TPM_HEAD_Genotype_Sex_Age_Cluster,
                   STRAIN = "DGRP-551",
                   SEX = "male",
                   GENE =III_Marker_Genes[i],
                   AGE1 ="50",
                   TOP_N_CLUSTERS = 12)
dev.off()
}

```



F2 before and after plot for 2 ages
```{r}
plot_gene_expression_before_after_2ages <- function(df, gene, genotype, sex, young_age, old_age,TOP_N_CLUSTERS) {
  library(ggplot2)
  library(dplyr)
  # Filter columns based on the provided inputs (genotype, sex, young and old ages)
  pattern_young <- paste0(genotype, "_", sex, "_", young_age)
  pattern_old <- paste0(genotype, "_", sex, "_", old_age)
  
  # Extract columns that match young and old age conditions
  young_cols <- grep(pattern_young, colnames(df), value = TRUE)
  old_cols <- grep(pattern_old, colnames(df), value = TRUE)
  
  # Ensure the same clusters are present in both ages
  clusters_young <- sub(paste0("_", young_age), "", young_cols)
  clusters_old <- sub(paste0("_", old_age), "", old_cols)
  
  # Find the top 7 clusters of a gene's expression in old age
  df_tmp <- df[gene,];
  df_tmp <- df_tmp[,old_cols]
  df_tmp <- as.matrix(df_tmp)
  df_tmp <- df_tmp[, order(df_tmp[1, ], decreasing = TRUE)];
  #select first 12 clusters to show
  df_tmp <- df_tmp[1:TOP_N_CLUSTERS]
  df_tmp <- as.matrix(df_tmp);
  df_tmp <- t(df_tmp);
  
  #renew variables of old_cols and clusters_old
  old_cols <- grep(pattern_old, colnames(df_tmp), value = TRUE)
  clusters_old <- sub(paste0("_", old_age), "", old_cols)
  
  #find common clusters between old and young clusters
  common_clusters <- intersect(clusters_young, clusters_old)
  
  # Subset the data for the common clusters
  young_cols_filtered <- young_cols[clusters_young %in% common_clusters]
  old_cols_filtered <- old_cols[clusters_old %in% common_clusters]
  
  # Get expression values for the gene in young and old conditions
  expression_young <- df[gene, young_cols_filtered]
  expression_old <- df[gene, old_cols_filtered]
  
  # Create a dataframe for plotting
  plot_data <- data.frame(
    Cluster = str_split(common_clusters,"_",3,simplify = T)[,3],
    Young_Age = log2(as.numeric(expression_young)+1),
    Old_Age = log2(as.numeric(expression_old)+1)
  )
   # Reshape the data into long format for ggplot input
  plot_data_long <- plot_data %>%
    pivot_longer(cols = c(Young_Age, Old_Age), names_to = "Age", values_to = "Expression") %>%
    mutate(Age = ifelse(Age == "Young_Age", "Young Age", "Old Age"))
  
  # Create the before-and-after plot 
  ggplot(plot_data_long, aes(x = Age, y = Expression, group = Cluster, shape = Cluster, color = Cluster)) +
    geom_point(size = 3) +
    geom_line() +
    theme_minimal() +
    labs(
      title = paste0("log2(TPM) change of ", gene, " in ", genotype," ",sex, " from ",young_age," to ",old_age, " in all clusters"),
      x = "Age Group",
      y = "Expression Level",
      shape = "Cluster",
      color = "Cluster"#it is better to use shapes, colors are indeed easily to plot, looks better, but very difficult to distinguish REMOVE 73 124
    )+
    scale_x_discrete(limits = c("Young Age", "Old Age"))+
    scale_shape_manual(values=c(33:72,74:123,125:127), breaks = waiver(), na.value = NA)#use ?points and check pch value options
}

# Example:

plot_gene_expression_before_after_2ages(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                                  gene = "Mtk",
                                  genotype = "DGRP-551",
                                  sex = "female",
                                  young_age = 3,
                                  old_age = 50,
                                  TOP_N_CLUSTERS = 7 
                                  )

plot_gene_expression_before_after_2ages(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                                  gene = "Mtk",
                                  genotype = "DGRP-551",
                                  sex = "female",
                                  young_age = 3,
                                  old_age = 30,
                                  TOP_N_CLUSTERS = 7
                                  )
```

Generate F2 for DGRP551 female/male target_ImmuneGenes and IMs
```{r}
#female_target_ImmuneGenes_50
for (i in 1:length(female_target_ImmuneGenes_50)) {
  tmp <- plot_gene_expression_before_after_2ages(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                                  gene = female_target_ImmuneGenes_50[i],
                                  genotype = "DGRP-551",
                                  sex = "female", 
                                  young_age = 3, 
                                  old_age = 50,
                                  TOP_N_CLUSTERS = 7
                                  );
  ggsave(tmp,device = "pdf",file= paste0("./results/Before_and_after_plot/female_target_ImmuneGenes_50_plot/Expression_change_DGRP-551_female_day3_day50_",female_target_ImmuneGenes_50[i],".pdf"),width = 9,height = 6)
}

#male_target_ImmuneGenes_50
for (i in 1:length(male_target_ImmuneGenes_50)) {
  tmp <- plot_gene_expression_before_after_2ages(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                                  gene = male_target_ImmuneGenes_50[i],
                                  genotype = "DGRP-551",
                                  sex = "male", 
                                  young_age = 3, 
                                  old_age = 50,
                                  TOP_N_CLUSTERS = 7
                                  );
  ggsave(tmp,device = "pdf",file= paste0("./results/Before_and_after_plot/male_target_ImmuneGenes_50_plot/Expression_change_DGRP-551_male_day3_day50_",male_target_ImmuneGenes_50[i],".pdf"),width = 9,height = 6)
}

#female IMs
for (i in 1:length(III_Marker_Genes)) {
  tmp <- plot_gene_expression_before_after_2ages(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                                  gene = III_Marker_Genes[i],
                                  genotype = "DGRP-551",
                                  sex = "female", 
                                  young_age = 3, 
                                  old_age = 50,
                                  TOP_N_CLUSTERS = 7
                                  );
  ggsave(tmp,device = "pdf",file= paste0("./results/Before_and_after_plot/female_IMs/Expression_change_DGRP-551_female_day3_day50_",III_Marker_Genes[i],".pdf"),width = 9,height = 6)
}

#male IMs
for (i in 1:length(III_Marker_Genes)) {
  tmp <- plot_gene_expression_before_after_2ages(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                                  gene = III_Marker_Genes[i],
                                  genotype = "DGRP-551",
                                  sex = "male", 
                                  young_age = 3, 
                                  old_age = 50,
                                  TOP_N_CLUSTERS = 7
                                  );
  ggsave(tmp,device = "pdf",file= paste0("./results/Before_and_after_plot/male_IMs/Expression_change_DGRP-551_male_day3_day50_",III_Marker_Genes[i],".pdf"),width = 9,height = 6)
}

```


F3 before and after plot for 4 ages (an extension of F2)
```{r}
plot_gene_expression_before_after_4ages <- function(df, gene, genotype, sex, first_age, second_age, third_age, fourth_age, TOP_N_CLUSTERS, cell_number_df) {
  library(ggplot2)
  library(dplyr)
  library(tidyr)
  
  # Generate the column pattern for the specific genotype, sex (age is filtered in later part of the function)
  pattern <- paste0(genotype, "_", sex, "_")
  
  # Filter columns matching the pattern
  selected_cols <- grep(pattern, colnames(df), value = TRUE)
  
  # Exclude unannotated clusters
  selected_cols <- selected_cols[!grepl("^[0-9]+$", sub(".*_", "", selected_cols))]
  
  
  # Subset the expression matrix
  filtered_df <- df[, selected_cols]
  filtered_df <- as.data.frame(filtered_df);
  
  # Filter columns based on metadata inputs
  pattern_first <- paste0(genotype, "_", sex, "_", first_age, "_")
  pattern_second <- paste0(genotype, "_", sex, "_", second_age, "_")
  pattern_third <- paste0(genotype, "_", sex, "_", third_age, "_")
  pattern_fourth <- paste0(genotype, "_", sex, "_", fourth_age, "_")
  
  first_cols <- grep(pattern_first, colnames(df), value = TRUE)
  second_cols <- grep(pattern_second, colnames(df), value = TRUE)
  third_cols <- grep(pattern_third, colnames(df), value = TRUE)
  fourth_cols <- grep(pattern_fourth, colnames(df), value = TRUE)
  
  # Ensure the same clusters are present
  clusters_first <- sub(paste0("_", first_age), "", first_cols)
  clusters_second <- sub(paste0("_", second_age), "", second_cols)
  clusters_third <- sub(paste0("_", third_age), "", third_cols)
  clusters_fourth <- sub(paste0("_", fourth_age), "", fourth_cols)
  
  # Identify top n clusters by expression in the fourth_age
  df_tmp <- filtered_df
  df_tmp <- as.data.frame(df_tmp)
  df_tmp <- df_tmp[,colnames(df_tmp)%in%fourth_cols]
  df_tmp <- df_tmp[rownames(df_tmp)==gene,]
  df_tmp <- as.matrix(df_tmp)
  df_tmp <- df_tmp[, order(df_tmp[1, ], decreasing = TRUE)]
  df_tmp <- df_tmp[1:TOP_N_CLUSTERS]
  df_tmp <- as.data.frame(df_tmp)
  
  # re-enter value for 1:TOP_N_CLUSTERS in clusters_fourth
  fourth_cols <- rownames(df_tmp)
  clusters_fourth <- sub(paste0("_", fourth_age), "", fourth_cols)
  
  # Find common clusters
  common_clusters <- Reduce(intersect, list(clusters_first, clusters_second, clusters_third, clusters_fourth))
  
  # Subset columns for common clusters
  first_cols_filtered <- first_cols[clusters_first %in% common_clusters]
  second_cols_filtered <- second_cols[clusters_second %in% common_clusters]
  third_cols_filtered <- third_cols[clusters_third %in% common_clusters]
  fourth_cols_filtered <- fourth_cols[clusters_fourth %in% common_clusters]
  
  # Prepare `cell_number_df` to include separate columns for metadata components
  cell_number_df <- as.data.frame(t(cell_number_df))
  colnames(cell_number_df) <- "Total_Cell_Count"
  cell_number_df$Metadata <- rownames(cell_number_df)
  cell_number_df <- cell_number_df %>%
    separate(Metadata, into = c("Genotype", "Sex", "Age", "Cluster"), sep = "_")
  
  # Calculate the total cell count across all ages for each Genotype-Sex-Cluster combination
  total_cell_counts <- cell_number_df %>%
    group_by(Genotype, Sex, Cluster) %>%
    summarise(Total_Cluster_Count_All_Ages = sum(Total_Cell_Count, na.rm = TRUE))
  
  # Join to calculate percentages
  cell_number_df <- cell_number_df %>%
    left_join(total_cell_counts, by = c("Genotype", "Sex", "Cluster")) %>%
    mutate(Cell_Percentage = Total_Cell_Count / Total_Cluster_Count_All_Ages * 100)
  
  # Prepare data for plotting
  expression_first <- df[gene, first_cols_filtered]
  expression_second <- df[gene, second_cols_filtered]
  expression_third <- df[gene, third_cols_filtered]
  expression_fourth <- df[gene, fourth_cols_filtered]
  
  plot_data <- data.frame(
    Cluster = sub(paste0("_", first_age), "", first_cols_filtered),
    First_Age = log2(as.numeric(expression_first) + 1),
    Second_Age = log2(as.numeric(expression_second) + 1),
    Third_Age = log2(as.numeric(expression_third) + 1),
    Fourth_Age = log2(as.numeric(expression_fourth) + 1)
  )
  
  # Pivot for ggplot
  plot_data_long <- plot_data %>%
    pivot_longer(cols = c(First_Age, Second_Age, Third_Age, Fourth_Age),
                 names_to = "Age", values_to = "Expression") %>%
    mutate(Age = factor(Age, levels = c("First_Age", "Second_Age", "Third_Age", "Fourth_Age"),
                        labels = c(first_age, second_age, third_age, fourth_age)));
  #change colnames for merge
  colnames(plot_data_long)[1] <- "Genotype_Sex_Cluster"
  plot_data_long$"Genotype_Sex_Cluster_Age" <- paste0(plot_data_long$"Genotype_Sex_Cluster","_",plot_data_long$Age)
  #change colnames for merge
  cell_number_df$"Genotype_Sex_Cluster_Age" <- paste0(cell_number_df$Genotype,"_",cell_number_df$Sex,"_",cell_number_df$Cluster,"_",cell_number_df$Age);
  #merge plot_data_long and cell_number_df
  plot_data_long <- merge(plot_data_long,cell_number_df,by="Genotype_Sex_Cluster_Age")
  plot_data_long <- plot_data_long[,-3];
  colnames(plot_data_long)[7] <- "Age"
  
  plot_data_long$Age <- factor(plot_data_long$Age, levels = c(first_age, second_age, third_age, fourth_age))
  
  # Plot
  ggplot(plot_data_long, aes(x = Age, y = Expression, group = Cluster, color = Cluster)) +
    geom_line(size = 1) +
    geom_point(aes(size = Cell_Percentage,stroke = Total_Cell_Count), shape = 21, fill = NA,alpha = 0.15) +
    scale_size_continuous(name = "Cell Percentage (%)") +
    #scale_color_discrete(name = "Cluster") +
    #scale_size_continuous(name = "Total Cell Count (Thickness)", range = c(0.5, 2)) + # Add legend for stroke
    theme_minimal() +
    labs(
      title = paste0("Multi-Age Plot for Gene: ", gene, " in ", genotype, " ", sex),
      x = "Age Group",
      y = "log2(TPM) Expression Level",
      color = "Cluster"
    ) +
    theme(legend.position = "right")
}

# Example:
plot_gene_expression_before_after_4ages(
  df=TPM_HEAD_Genotype_Sex_Age_Cluster,
  gene = "Dro",
  genotype = "DGRP-551",
  sex = "female",
  first_age = 3,
  second_age = 15,
  third_age = 30,
  fourth_age = 50,
  TOP_N_CLUSTERS = 7,
  cell_number_df = cell_number_Genotype_Sex_Age_Cluster
  )


```


Generate F3 for DGRP551 female/male target_ImmuneGenes and IMs
```{r}
#female_target_ImmuneGenes_50
for (i in 1:length(female_target_ImmuneGenes_50)) {
  tmp <- plot_gene_expression_before_after_4ages(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                                  gene = female_target_ImmuneGenes_50[i],
                                  genotype = "DGRP-551",
                                  sex = "female",
                                  first_age = 3,
                                  second_age = 15,
                                  third_age = 30,
                                  fourth_age = 50,
                                  TOP_N_CLUSTERS = 7,
                                  cell_number_df = cell_number_Genotype_Sex_Age_Cluster
                                  );
  ggsave(tmp,device = "pdf",file= paste0("./results/Before_and_after_plot/female_target_ImmuneGenes_50_plot/Expression_change_DGRP-551_female_day3_day_15_day30_day50_",female_target_ImmuneGenes_50[i],".pdf"),width = 9,height = 6)
}

#male_target_ImmuneGenes_50
for (i in 1:length(male_target_ImmuneGenes_50)) {
  tmp <- plot_gene_expression_before_after_4ages(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                                  gene = male_target_ImmuneGenes_50[i],
                                  genotype = "DGRP-551",
                                  sex = "male", 
                                  first_age = 3,
                                  second_age = 15,
                                  third_age = 30,
                                  fourth_age = 50,
                                  TOP_N_CLUSTERS = 7,
                                  cell_number_df = cell_number_Genotype_Sex_Age_Cluster
                                  );
  ggsave(tmp,device = "pdf",file= paste0("./results/Before_and_after_plot/male_target_ImmuneGenes_50_plot/Expression_change_DGRP-551_male_day3_day_15_day30_day50_",male_target_ImmuneGenes_50[i],".pdf"),width = 9,height = 6)
}

#female IMs
for (i in 1:length(III_Marker_Genes)) {
  tmp <- plot_gene_expression_before_after_4ages(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                                  gene = III_Marker_Genes[i],
                                  genotype = "DGRP-551",
                                  sex = "female", 
                                  first_age = 3,
                                  second_age = 15,
                                  third_age = 30,
                                  fourth_age = 50,
                                  TOP_N_CLUSTERS = 7,
                                  cell_number_df = cell_number_Genotype_Sex_Age_Cluster
                                  );
  ggsave(tmp,device = "pdf",file= paste0("./results/Before_and_after_plot/female_IMs/Expression_change_DGRP-551_female_day3_day_15_day30_day50_",III_Marker_Genes[i],".pdf"),width = 9,height = 6)
}

#male IMs
for (i in 1:length(III_Marker_Genes)) {
  tmp <- plot_gene_expression_before_after_4ages(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                                  gene = III_Marker_Genes[i],
                                  genotype = "DGRP-551",
                                  sex = "male", 
                                  first_age = 3,
                                  second_age = 15,
                                  third_age = 30,
                                  fourth_age = 50,
                                  TOP_N_CLUSTERS = 7,
                                  cell_number_df = cell_number_Genotype_Sex_Age_Cluster
                                  );
  ggsave(tmp,device = "pdf",file= paste0("./results/Before_and_after_plot/male_IMs/Expression_change_DGRP-551_male_day3_day_15_day30_day50_",III_Marker_Genes[i],".pdf"),width = 9,height = 6)
}

```

F4 Heat map function (Does not include unannotated clusters)
```{r}
# note: the function currently generates absolute expression for informative purpose; normalization to columns or rows renders heat map cells indistinguishable 

plot_gene_expression_heatmap <- function(df, genes, genotype, sex, age) {
  # Load necessary libraries
  library(ggplot2)
  library(ComplexHeatmap)
  library(dplyr)
  library(tidyr)
  library(colorRamp2)
  
  # Generate the column pattern for the specific genotype, sex, and age
  pattern <- paste0(genotype, "_", sex, "_", age, "_")
  
  # Filter columns matching the pattern
  selected_cols <- grep(pattern, colnames(df), value = TRUE)
  
  # Exclude unannotated clusters
  selected_cols <- selected_cols[!grepl("^[0-9]+$", sub(".*_", "", selected_cols))]
  
  # Subset the expression matrix
  filtered_df <- df[rownames(df) %in% genes, selected_cols, drop = FALSE]
  
  # Check if the filtered_df has valid data
  if (nrow(filtered_df) == 0 || ncol(filtered_df) == 0) {
    stop("No matching data found for the specified parameters.")
  }
  
  # Prepare a matrix for the heatmap
  heatmap_matrix <- as.matrix(filtered_df)
  
  # log2 transformation (to make the color span less)
  heatmap_matrix <- log2(heatmap_matrix+1)
  
  # Define a color scale for the heatmap
  col_fun <- colorRamp2(
    breaks = c(min(heatmap_matrix), median(heatmap_matrix), max(heatmap_matrix)),
    colors = c("#008080", "#FFFFFF", "#FFD700")
  )
  
  # Limit colnames of heatmap_matrix to only clusters
  colnames(heatmap_matrix) <- 
    str_split(colnames(as.data.frame(heatmap_matrix)),pattern = "_",n = 4,simplify =T)[,4]
  
  # Generate the heatmap
  heatmap <- Heatmap(
    heatmap_matrix,
    name = "log2(Expression)",
    col = col_fun,
    row_names_side = "left",
    column_names_side = "bottom",
    cluster_rows = F, # similarity clusters on rows (genes)
    cluster_columns = F, # similarity clusters on columns (clusters); enabled
    show_row_names = TRUE,
    show_column_names = TRUE,
    row_title = "Genes",
    column_title = paste("Clusters (", genotype, ", ", sex, ", Age: ", age, ")", sep = ""),
    cell_fun = function(j, i, x, y, width, height, fill) {
      grid.rect(x, y, width, height, gp = gpar(col = "black", lwd = 1))  # Add black border
      grid.rect(x, y, width, height, gp = gpar(fill = fill, col = NA))    # Preserve fill color
    }
  )
  
  # Draw the heatmap
  draw(heatmap)
}



# usage example

# DGRP-551
plot_gene_expression_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                             genes=female_target_ImmuneGenes_50,
                             genotype="DGRP-551",
                             sex="female", 
                             age="3")

plot_gene_expression_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                             genes=female_target_ImmuneGenes_50,
                             genotype="DGRP-551",
                             sex="female", 
                             age="50")

# plot_gene_expression_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
#                              genes=male_target_ImmuneGenes_50,
#                              genotype="DGRP-551",
#                              sex="male", 
#                              age="30")
# 
# plot_gene_expression_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
#                              genes=male_target_ImmuneGenes_50,
#                              genotype="DGRP-551",
#                              sex="male", 
#                              age="50")
# # w1118
# plot_gene_expression_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
#                              genes=female_target_ImmuneGenes_50,
#                              genotype="w1118",
#                              sex="female", 
#                              age="30")
# 
# plot_gene_expression_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
#                              genes=male_target_ImmuneGenes_50,
#                              genotype="w1118",
#                              sex="male", 
#                              age="30")
```




F5  plot log2((old expression + 1) / (young expression + 1)) for common clusters

```{r}

plot_age_ratio_heatmap <- function(df, genes, genotype, sex, young_age, old_age) {
  library(ComplexHeatmap)
library(circlize)
library(viridis) # heatmap color
  # Generate column patterns for both ages
  young_pattern <- paste0(genotype, "_", sex, "_", young_age, "_")
  old_pattern <- paste0(genotype, "_", sex, "_", old_age, "_")
  
  # Extract relevant column names
  young_cols <- grep(young_pattern, colnames(df), value = TRUE)
  old_cols <- grep(old_pattern, colnames(df), value = TRUE)
  
  # Extract cluster names
  young_clusters <- sub(paste0(".*_", young_age, "_"), "", young_cols)
  old_clusters <- sub(paste0(".*_", old_age, "_"), "", old_cols)
  
  # Exclude unannotated clusters (numeric cluster names)
  valid_young <- !grepl("^[0-9]+$", young_clusters)
  valid_old <- !grepl("^[0-9]+$", old_clusters)
  
  young_cols <- young_cols[valid_young]
  old_cols <- old_cols[valid_old]
  
  young_clusters <- young_clusters[valid_young]
  old_clusters <- old_clusters[valid_old]
  
  # Find common clusters
  common_clusters <- intersect(young_clusters, old_clusters)
  
  if (length(common_clusters) == 0) {
    stop("No common annotated clusters found between the two ages.")
  }
  
  # Subset data for common clusters
  young_filtered <- df[rownames(df) %in% genes, young_cols[young_clusters %in% common_clusters], drop = FALSE]
  old_filtered <- df[rownames(df) %in% genes, old_cols[old_clusters %in% common_clusters], drop = FALSE]
  
  # Rename columns to cluster names for alignment
  colnames(young_filtered) <- common_clusters
  colnames(old_filtered) <- common_clusters
  
  # Compute log2 ratio (log2(old + 1) / log2(young + 1))
  log2_ratio <- log2((old_filtered + 1) / (young_filtered + 1))
  
  # Define color scale (matching your figure: blue < white < yellow)
  #col_fun <- colorRamp2(c(-15, 0, 15), c("violetred","#008080", "#FFD700"))
  #col_fun <- colorRamp2(c(-5,0,5), c("navy","aquamarine4", "darkorange"))
  #col_fun <- colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))  # Adjust values as needed
  col_fun <- colorRamp2(
  c(-3, -1, 1, 3),  # Adjust fold-change values for transitions
  c("darkblue", "lightblue", "lightsalmon", "darkred")  # 4-color gradient
)

  
  # Plot heatmap
  Heatmap(as.matrix(log2_ratio),
          name = "log2 Fold Change",
          col = col_fun,
          cluster_rows = FALSE,
          cluster_columns = T,
          show_row_names = TRUE,  # Genes on the left
          show_column_names = TRUE, # Cell populations
          column_names_rot = 90,  # Vertical column names
          row_names_side = "left",
          border = TRUE, # Adds grid lines
          column_title = paste0("Log2 Fold Change (", old_age, "/", young_age, ")"," of ",genotype," ", sex),
          row_title = "Genes",
          column_names_gp = gpar(fontsize = 10),  # Adjust font size
          heatmap_legend_param = list(title = "Log2 FC"),
          cell_fun = function(j, i, x, y, width, height, fill) {
      grid.rect(x, y, width, height, gp = gpar(col = "black", lwd = 1))  # Add black border
      grid.rect(x, y, width, height, gp = gpar(fill = fill, col = NA))    # Preserve fill color
    }
  )
}


# Example
plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                       genes=female_target_ImmuneGenes_50,
                       genotype="DGRP-551",
                       sex="female",
                       young_age="3",
                       old_age="50")
# 
# plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
#                        genes=female_target_ImmuneGenes_30,
#                        genotype="DGRP-551",
#                        sex="female",
#                        young_age="3",
#                        old_age="30")

plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                       genes=male_target_ImmuneGenes_50,
                       genotype="DGRP-551",
                       sex="male",
                       young_age="3",
                       old_age="50")
# 
# plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
#                        genes=male_target_ImmuneGenes_30,
#                        genotype="DGRP-551",
#                        sex="male",
#                        young_age="3",
#                        old_age="30")
#######
# plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
#                        genes=III_Marker_Genes,
#                        genotype="DGRP-551",
#                        sex="female",
#                        young_age="3",
#                        old_age="50")
# plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
#                        genes=III_Marker_Genes,
#                        genotype="DGRP-551",
#                        sex="male",
#                        young_age="3",
#                        old_age="50")
# plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
#                        genes=III_Marker_Genes,
#                        genotype="DGRP-551",
#                        sex="female",
#                        young_age="3",
#                        old_age="30")
# plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
#                        genes=III_Marker_Genes,
#                        genotype="DGRP-551",
#                        sex="male",
#                        young_age="3",
#                        old_age="30")

```
fold change plot of all brain IMs in female and male DGRP-551
```{r}
############### Brain IMs #################
# all brain IMs in female DGRP-551
plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                       genes=unique(c(female_target_ImmuneGenes_50,male_target_ImmuneGenes_50)),
                       genotype="DGRP-551",
                       sex="female",
                       young_age="3",
                       old_age="50")

# all brain IMs in male DGRP-551
plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                       genes=unique(c(female_target_ImmuneGenes_50,male_target_ImmuneGenes_50)),
                       genotype="DGRP-551",
                       sex="male",
                       young_age="3",
                       old_age="50")

##################### WB IMs ##############

# WB IMs in female DGRP-551
plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                       genes=unique(c(III_Marker_Genes)),
                       genotype="DGRP-551",
                       sex="female",
                       young_age="3",
                       old_age="50")

# WB IMs in male DGRP-551
plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                       genes=unique(c(III_Marker_Genes)),
                       genotype="DGRP-551",
                       sex="male",
                       young_age="3",
                       old_age="50")


##################### Brain IMs + WB IMs #######################
# all brain IMs and WB IMs in female DGRP-551
plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                       genes=unique(c(female_target_ImmuneGenes_50,male_target_ImmuneGenes_50,III_Marker_Genes)),
                       genotype="DGRP-551",
                       sex="female",
                       young_age="3",
                       old_age="50")

# all brain IMs and WB IMs in male DGRP-551
plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                       genes=unique(c(female_target_ImmuneGenes_50,male_target_ImmuneGenes_50,III_Marker_Genes)),
                       genotype="DGRP-551",
                       sex="male",
                       young_age="3",
                       old_age="50")




```
fold change day 50/ day 3 TPM plot of all immune genes
```{r}
# DGRP-551 female
pdf("./results/Heatmap/GO_ImmuneDef_day50_day3_DGRP-551_female.pdf",
    width = 10,
    height = 70)
plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                       genes=unique(GO_ImmuneDef$V1),
                       genotype="DGRP-551",
                       sex="female",
                       young_age="3",
                       old_age="50")
dev.off()

# DGRP-551 male
pdf("./results/Heatmap/GO_ImmuneDef_day50_day3_DGRP-551_male.pdf",
    width = 10,
    height = 70)
plot_age_ratio_heatmap(df=TPM_HEAD_Genotype_Sex_Age_Cluster,
                       genes=unique(GO_ImmuneDef$V1),
                       genotype="DGRP-551",
                       sex="male",
                       young_age="3",
                       old_age="50")
dev.off()
```

F6 plot_selected_clusters_age_ratio_heatmap for specific cell types
```{r}
plot_selected_clusters_age_ratio_heatmap <- function(df, genes, genotype, sex, young_age, old_age, clusters_to_plot = NULL) {
  library(ComplexHeatmap)
  library(circlize)
  library(viridis)

  # Generate column patterns for both ages
  young_pattern <- paste0(genotype, "_", sex, "_", young_age, "_")
  old_pattern <- paste0(genotype, "_", sex, "_", old_age, "_")
  
  # Extract relevant column names
  young_cols <- grep(young_pattern, colnames(df), value = TRUE)
  old_cols <- grep(old_pattern, colnames(df), value = TRUE)
  
  # Extract cluster names
  young_clusters <- sub(paste0(".*_", young_age, "_"), "", young_cols)
  old_clusters <- sub(paste0(".*_", old_age, "_"), "", old_cols)
  
  # Exclude unannotated clusters (numeric only names)
  valid_young <- !grepl("^[0-9]+$", young_clusters)
  valid_old <- !grepl("^[0-9]+$", old_clusters)
  young_cols <- young_cols[valid_young]
  old_cols <- old_cols[valid_old]
  young_clusters <- young_clusters[valid_young]
  old_clusters <- old_clusters[valid_old]
  
  # Find common annotated clusters
  common_clusters <- intersect(young_clusters, old_clusters)

  # If user specifies clusters, filter them
  if (!is.null(clusters_to_plot)) {
    common_clusters <- intersect(common_clusters, clusters_to_plot)
  }

  if (length(common_clusters) == 0) {
    stop("No matching annotated clusters found between the two ages based on your selection.")
  }
  
  # Subset data for common clusters
  young_filtered <- df[rownames(df) %in% genes, young_cols[young_clusters %in% common_clusters], drop = FALSE]
  old_filtered <- df[rownames(df) %in% genes, old_cols[old_clusters %in% common_clusters], drop = FALSE]

  # Rename columns to only the cluster names
  colnames(young_filtered) <- sub(paste0(".*_", young_age, "_"), "", colnames(young_filtered))
  colnames(old_filtered) <- sub(paste0(".*_", old_age, "_"), "", colnames(old_filtered))
  
  # Make sure columns are in the same order
  common_clusters_sorted <- sort(common_clusters)
  young_filtered <- young_filtered[, common_clusters_sorted, drop = FALSE]
  old_filtered <- old_filtered[, common_clusters_sorted, drop = FALSE]

  # Compute log2 ratio
  log2_ratio <- log2((old_filtered + 1) / (young_filtered + 1))
  
  # Define 4-color scale (no white, smooth red-blue)
  col_fun <- colorRamp2(
    c(-3, -1, 1, 3),
    c("darkblue", "lightblue", "lightsalmon", "darkred")
  )

  # Plot heatmap
  Heatmap(
    as.matrix(log2_ratio),
    name = "log2 Fold Change",
    col = col_fun,
    cluster_rows = FALSE,
    cluster_columns = TRUE,
    show_row_names = TRUE,
    show_column_names = TRUE,
    column_names_rot = 90,
    row_names_side = "left",
    border = TRUE,
    column_title = paste0("Log2 Fold Change (", old_age, "/", young_age, ") of ", genotype, " ", sex),
    row_title = "Genes",
    column_names_gp = gpar(fontsize = 10),
    heatmap_legend_param = list(title = "Log2 FC"),
    cell_fun = function(j, i, x, y, width, height, fill) {
      grid.rect(x, y, width, height, gp = gpar(col = "black", lwd = 1))  # Draw grid
      grid.rect(x, y, width, height, gp = gpar(fill = fill, col = NA))   # Preserve cell color
    }
  )
}



# example

# all brain IMs in female DGRP-551
plot_selected_clusters_age_ratio_heatmap(
  df = TPM_HEAD_Genotype_Sex_Age_Cluster,
  genes = unique(c(female_target_ImmuneGenes_50,male_target_ImmuneGenes_50)),
  genotype = "DGRP-551",
  sex = "female",
  young_age = "3",
  old_age = "50",
  clusters_to_plot = c("Plasmatocytes","MBON","Astrocyte-like","Dopaminergic")  # input specific clusters; check annotated_clusters for more info
)


# all brain IMs in male DGRP-551
plot_selected_clusters_age_ratio_heatmap(
  df = TPM_HEAD_Genotype_Sex_Age_Cluster,
  genes = unique(c(female_target_ImmuneGenes_50,male_target_ImmuneGenes_50)),
  genotype = "DGRP-551",
  sex = "male",
  young_age = "3",
  old_age = "50",
  clusters_to_plot = c("Plasmatocytes","MBON","Astrocyte-like","Dopaminergic")  # input specific clusters; check annotated_clusters for more info
)




```



venn grey and show gene names in sets of intersection
```{r}
# Load required packages
library(ggplot2)
library(ggvenn)

# Define three gene lists
gene_list1 <- III_Marker_Genes
gene_list2 <- female_target_ImmuneGenes_50
gene_list3 <- male_target_ImmuneGenes_50

# Create a named list of gene sets
gene_sets <- list(
  "WB IMs" = gene_list1,
  "Female Brain IMs Day 50" = gene_list2,
  "Male Brain IMs Day 50" = gene_list3
)

# Compute intersections
intersect_12 <- intersect(gene_list1, gene_list2)
intersect_13 <- intersect(gene_list1, gene_list3)
intersect_23 <- intersect(gene_list2, gene_list3)
intersect_123 <- Reduce(intersect, list(gene_list1, gene_list2, gene_list3))

# Generate Venn Diagram
p <- ggvenn(gene_sets, fill_color = c("grey", "grey", "grey"), stroke_size = 0.5, set_name_size = 4,fill_alpha = 0.5,text_size = 5,show_percentage = F)

# Add annotations for intersecting genes
p + 
  annotate("text", x = 0, y = 0.5, label = paste(setdiff(intersect_12, intersect_123), collapse = ", "), size = 4) +  # Intersection of List1 and List2 excluding intersect_123
  annotate("text", x = 0, y = -1, label = paste(setdiff(gene_list3, intersect_123), collapse = ", "), size = 4) +   # gene_list3 excluding intersect_123
  annotate("text", x = 0, y = 0, label = paste(intersect_123, collapse = ", "), size = 4) +    # Intersection of all three lists
  annotate("text", x = 0.8, y = 0.4, label = paste(setdiff(gene_list2, intersect_12), collapse = ", "), size = 4)     # gene_list3 excluding intersect_12
```









Day 50 Venn Padj0.01
```{r}
library(ggVennDiagram)

# Create a list for the Venn diagram
Venn_gene_lists <- list(
  'Whole Body IMs' = III_Marker_Genes,
  'Female Brain IMs' = female_target_ImmuneGenes_50,
  'Male Brain IMs' = male_target_ImmuneGenes_50
)


# Plot Venn diagram with adjusted margins
pdf("./results/Venn/Venn_IMs_WB_F_M_50.pdf",
    width = 9,
    height = 8)
ggVennDiagram(Venn_gene_lists, label_alpha = 0) + 
  scale_fill_gradient(low = "lightblue", high = "darkblue") +  # Fill gradient
  labs(title = "Venn Diagram of Gene Lists") +  # Add title
  theme(
    plot.title = element_text(size = 18, face = "bold", hjust = 0.5),  # Center title
    plot.margin = margin(20, 50, 20, 50),  # Increase margins (top, right, bottom, left)
    text = element_text(size = 14)  # Increase overall text size
  ) +
  coord_cartesian(clip = "off")  # Prevent text from being cut off
dev.off()
```


Day 30 Venn Padj0.01
```{r}
library(ggVennDiagram)

# Create a list for the Venn diagram
Venn_gene_lists <- list(
  'Whole Body IMs' = III_Marker_Genes,
  'Female Brain IMs' = female_target_ImmuneGenes_30,
  'Male Brain IMs' = male_target_ImmuneGenes_30
)


# Plot Venn diagram with adjusted margins
pdf("./results/Venn/Venn_IMs_WB_F_M_30.pdf",
    width = 9,
    height = 8)
ggVennDiagram(Venn_gene_lists, label_alpha = 0) + 
  scale_fill_gradient(low = "lightblue", high = "darkblue") +  # Fill gradient
  labs(title = "Venn Diagram of Gene Lists") +  # Add title
  theme(
    plot.title = element_text(size = 18, face = "bold", hjust = 0.5),  # Center title
    plot.margin = margin(20, 50, 20, 50),  # Increase margins (top, right, bottom, left)
    text = element_text(size = 14)  # Increase overall text size
  ) +
  coord_cartesian(clip = "off")  # Prevent text from being cut off
dev.off()
```


save work image
```{r}
save.image("./GSE107451_head_ALL_57k_clean_padj0.01.RData")
```

