---
title: "E-MTAB-8698_EMBOJ_AuthorNI"
output: html_notebook
---
#refer to:
Cattenoz PB, Sakr R, Pavlidaki A, Delaporte C, Riba A, Molina N, Hariharan N, Mukherjee T, Giangrande A. Temporal specificity and heterogeneity of Drosophila immune cells. EMBO J. 2020 Jun 17;39(12):e104486. doi: 10.15252/embj.2020104486. Epub 2020 Mar 12. PMID: 32162708; PMCID: PMC7298292.

#Objectives:
## 1. explore in which hemocyte clusters are III markers expressed.
## 2. determine if III markers are differentially expressed in certain clusters than others
## 3. TBA

#Important files***:
## 1. E-MTAB-8698.sdrf.txt meta information for seurat object
## 2. filtered_feature_bc_matrix_NI_RRCZ22, unzipped folder that contains control (Non_Infected) group of hemocytes scRNAseq data
*** These files can be downloaded at https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E%E2%80%90MTAB%E2%80%908698



set wk dir and library package
```{r}

#load("./EMBOJ_hemo_AuthorNI.RData")
setwd(getwd())

library(Seurat)
library(ggplot2)
library(cowplot)
library(Matrix)
library(dplyr)
library(ggsci)
library(data.table)
library(tidyverse)
library(patchwork)
library(clustree)
library(remotes)
```

#import seurat data and author's annotation
```{r}

Author_annotation <- data.table::fread(
  "./NI_cell_cluster_ID.txt"
  )

Author_annotation <- as.data.frame(Author_annotation)

NI <- Seurat::Read10X("./filtered_feature_bc_matrix/")

```

#creat a obj of III markers
```{r}
III_Markers <- c("SPH93",
                                                                          "CecA1",
                                                                          "CecA2",
                                                                          "AttA",
                                                                          "AttB",
                                                                          "AttC",
                                                                          "DptA",
                                                                          "DptB",
                                                                          "Dro",
                                                                          "Mtk",
                                                                          "IM18",
                                                                          "edin")
```

#create seurat OBJ
```{r}

NI <- CreateSeuratObject(counts = NI,
                           project = "Non_Infected_RRCZ22", 
                           min.cells = 0,  
                           min.features = 0)
```

#first check if author's annotation is the same as meta.data slot in SOBJ
```{r}
View(NI@meta.data)

#add "-1" rownames to change cell ID in Author_annotation to match meta.data slot
Author_annotation$Cell_ID <- paste0(Author_annotation$Cell_ID,"-1")

#check if author's annotation has the same number of cells as meta.data slot in SOBJ
nrow(NI@meta.data)
#7606
nrow(Author_annotation)
#7517

#they are not the same
```

#subset SOBJ to match author's annotation
```{r}
#first change cell ID in RNA slot to match meta.data slot
# 
# rownames(NI@assays[["RNA"]]@cells) <- rownames(NI@meta.data)
# rownames(NI@assays[["RNA"]]@cells)

table(rownames(NI@meta.data)%in%Author_annotation$Cell_ID)

NI2 <- NI[,rownames(NI@meta.data)%in%Author_annotation$Cell_ID]

# subset function is problematic with SeuratV5, use R fundamental instead
# subset(NI,rownames(NI@meta.data)==Author_annotation$Cell_ID)


#check if NI2 has same cell number of Author_annotation
nrow(NI2@meta.data)
#7517
nrow(Author_annotation)
#7517
```

#merge author's annotation with NI2 meta.data slot
```{r}
View(NI2@meta.data)
table(rownames(NI2@meta.data)%in%Author_annotation$Cell_ID)
# TRUE 
# 7517 

NI2@meta.data$'annotation' <- Author_annotation$Cluster_ID
```


#change active idents into author's annotation
```{r}
Idents(NI2) <- NI2@meta.data$annotation
Idents(NI2)
table(Idents(NI2))
```


#run standard seurat pipeline
```{r}
# run standard anlaysis workflow
NI2 <- NormalizeData(NI2)
NI2 <- FindVariableFeatures(NI2)
NI2 <- ScaleData(NI2)

# min(nrow(NI2), ncol(NI2))
NI2 <- RunPCA(NI2,npcs = 50,verbose = F)

#see turning point
ElbowPlot(NI2, ndims=30, reduction="pca")


NI2 <- FindNeighbors(NI2, dims = 1:20, reduction = "pca", verbose=TRUE)

NI2 <- FindClusters(NI2, 
                    cluster.name = paste0("NI_clusters",
                                          seq(0.6,0.8,by=0.05)),
                    resolution = seq(0.6,0.8,by=0.05))

#see resolution
pdf(paste0("./cluster_tree.pdf"),width = 9,height = 6)
clustree(NI2,prefix = "NI_clusters")
dev.off()
```

#Run UMAP and tSNE dim reduction
```{r}
#Umap
NI3 <- RunUMAP(NI2, dims = 1:20, reduction = "pca", reduction.name = "UMAP_NI",verbose=F)

colnames(NI3@reductions[["UMAP_NI"]]@cell.embeddings) <- c('UMAP_1', 'UMAP_2')

DimPlot(NI3, reduction = "UMAP_NI", group.by = "annotation")

DimPlot(NI3, reduction = "UMAP_NI")

#tsne
NI3 <- RunTSNE(NI3, dims = 1:20, reduction = "pca", reduction.name = "tSNE_NI",verbose=F)

colnames(NI3@reductions[["tSNE_NI"]]@cell.embeddings) <- c('tSNE_1', 'tSNE_2')


DimPlot(NI3, reduction = "tSNE_NI", group.by = "annotation")

```

#save Dimplot
```{r}
#tSNE Dimplot
ggsave(
  DimPlot(NI3, reduction = "tSNE_NI", group.by = "annotation",label = T,label.size = 3),
       device = "pdf",file= paste0("./Dimplot/Dimplot_NI_tSNE.pdf"),
       width = 9,
       height = 6
       )

#UMAP Dimplot
ggsave(
  DimPlot(NI3, reduction = "UMAP_NI", group.by = "annotation",label = T,label.size = 3),
       device = "pdf",file= paste0("./Dimplot/Dimplot_NI_UMAP.pdf"),
       width = 9,
       height = 6
       )


```


#find marker genes
#For clarification, check https://www.biostars.org/p/409790/
```{r}

all.markers <- FindAllMarkers(NI3, test.use = 'wilcox', only.pos = TRUE, min.pct = 0.10, logfc.threshold = 0.25)

```

#extract significant markers and topmarkers in each cluster
```{r}

marker.sig <- all.markers %>%
    filter(p_val_adj < 0.05)

topmarker <- marker.sig %>% group_by(cluster) %>% 
    top_n(n = 5, wt = avg_log2FC)

```

#plot heatmap and fitplot of significant markers and topmarkers in each cluster
```{r}

DoHeatmap(NI3, features = topmarker$gene,size = 2) +
    theme(legend.position = "none", 
          axis.text.y = element_text(size = 6))

DotPlot(NI3, features = unique(topmarker$gene))
```

#plot the dotplot of III markers
```{r}
DotPlot(NI3, features = III_Markers)
```

# save Dotplot
```{r}
ggsave(DotPlot(NI3, features = III_Markers),
       device = "pdf",file= paste0("./Dotplot/Dotplot_NI_III_Markers.pdf"),
       width = 9.8,
       height = 6
       )

ggsave(DotPlot(NI3, features = unique(topmarker$gene))+
     theme(axis.text.x = element_text(angle = 90, size = 12), legend.text = element_text(size = 12), legend.title = element_text(size = 12)),
       device = "pdf",file= paste0("./Dotplot/Dotplot_NI_TopMarkers.pdf"),
       width = 18,
       height = 12
       )
```


#plot the featureplot of III markers and look at their distribution in different cells
```{r}
#feature plot based on tSNE reduction
for (i in 1:length(III_Markers)) {
  tmp <- Seurat::FeaturePlot(NI3, features = III_Markers[i],  cols = c("grey","darkred"), reduction = "tSNE_NI",keep.scale = "feature")
  ggsave(tmp,device = "pdf",file= paste0("./Featureplot/Featureplot_NI_tSNE_",III_Markers[i],".pdf"),width = 9,height = 6)
}

#feature plot based on UMAP reduction
for (i in 1:length(III_Markers)) {
  tmp <- Seurat::FeaturePlot(NI3, features = III_Markers[i],  cols = c("grey","darkred"), reduction = "UMAP_NI",keep.scale = "feature")
  ggsave(tmp,device = "pdf",file= paste0("./Featureplot/Featureplot_NI_UMAP_",III_Markers[i],".pdf"),width = 9,height = 6)
}
```



#########################
#save wk image
```{r}
save.image("./EMBOJ_hemo_AuthorNI.RData")
```


